---
title: 'A short-read *de novo* transcriptome construction pipeline optimized by long reads reveals novel developmentally regulated gene isoforms and disease targets in hundreds of eye samples'
author:
  - Vinay S. Swamy:
      institute: bg
  - Temesgen D. Fufa:
      institute: mgog
  - Robert B. Hufnagel:
      institute: mgog
  - David M. McGaughey:
      institute: bg
      correspondence: "yes"
      email: mcgaugheyd@mail.nih.gov
institute:
  - bg: Bioinformatics Group, Ophthalmic Genetics & Visual Function Branch, National Eye Institute, Institutes of Health
  - mgog: Medical Genetics and Ophthalmic Genomics Unit, National Eye Institute, National Institutes of Health
bibliography: ocular_txome_citations.bib
csl: elife.csl
output:
 word_document:
  reference_docx: reference_doc_v1.docx
  keep_md: yes
  pandoc_args:
    - '--lua-filter=scholarly-metadata.lua'
    - '--lua-filter=author-info-blocks.lua'
 html_document:
  df_print: paged
 pdf_document:
  df_print: kable
---

```{r setup, include=FALSE,echo=FALSE, cache = FALSE}
require("knitr")
## setting working directory
#root = '/Users/mcgaugheyd/git/ocular_transcriptomes_paper/'
#opts_knit$set(root.dir = "~/git/ocular_transcriptomes_paper/")
#opts_knit$set(root.dir = root)

#redoc cmd : dedoc(docx, wrap = 9999999, overwrite = T, verbose = T, track_changes = "criticmarkup")
opts_knit$set(root.dir = '/data/swamyvs/ocular_transcriptomes_paper/')
opts_chunk$set(echo = F, message = F, warning = F, include = F, dpi = 300)
```

```{r prep, include=FALSE,echo=FALSE}
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(ComplexUpset)
library(ggrepel)
library(ggpubr)
library(patchwork)
library(knitr)
library(grid)
library(png)
library(glue)
library(ggplotify)
library(scattermore)
library(flextable)
sample_table <- read_tsv('sampleTableFullV3.tsv') %>% filter(subtissue != 'synth')
load('clean_data/rdata/tissue_to_colors.Rdata')
load('clean_data/rdata/core_tight.Rdata')
tissue_color_mapping_df <- bind_rows(tissue_color_mapping_df, tibble(body_location=c('Brain(avg)', 'Body(avg)'), color=c('orange','yellow')))
load('clean_data/rdata/transcriptome_pipeline_stats.Rdata')
subtissue_to_bodyloc <- sample_table %>% select(subtissue, body_location) %>% distinct
core_tight <- core_tight %>% as_tibble %>% select(sample=sample_accession, study=study_accession) %>% distinct
load('clean_data/rdata/paper_numbers.Rdata')
load('clean_data/rdata/longread_summary.Rdata')
load('clean_data/rdata/buildResultsSummary.Rdata')
load('clean_data/rdata/novel_isoforms.Rdata')
#load('clean_data/rdata/novel_loci_fig.Rdata')
#load('clean_data/rdata/vep_results.Rdata')
load('clean_data/rdata/fetal_retina_diffexp_results.Rdata')
load('clean_data/rdata/CAGE_poly_dist.Rdata')
load('clean_data/rdata/intron_variant_analysis.Rdata')
color_list<- novel_transcripts_per_tissue$color
names(color_list) <- novel_transcripts_per_tissue$body_location
eye_tissues <- c("RPE_Fetal.Tissue", "Retina_Adult.Tissue", "RPE_Adult.Tissue", "Cornea_Adult.Tissue", "Cornea_Fetal.Tissue", "Retina_Fetal.Tissue")
subtissues <- unique(sample_table$subtissue)
body_tissues <- sample_table %>% filter(!subtissue%in% eye_tissues) %>% pull(subtissue) %>% unique
```

# Abstract

|              *De novo* transcriptome construction from short-read RNA-seq is a common method for reconstructing previously annotated and novel mRNA transcripts within a given sample. However, the precision of the built transcriptome is unclear as it is difficult to obtain a ground-truth measure of transcript expression. With advances in third generation sequencing, full length transcripts of whole transcriptomes can be accurately sequenced to generate a ground-truth transcriptome— but it is significantly more expensive than short-read sequencing. We generated long-read PacBio and short-read Illumina RNA sequencing data from an induced pluripotent stem cell- derived retinal pigmented epithelium (iPSC-RPE) cell line. We use the long-read data to identify simple but powerful metrics for assesing *de novo* transcriptome construction and to optimize a short-read based *de novo* transcriptome construction pipeline. We then apply this this pipeline to construct transcriptomes for `r toString(NUM_EYE_SAMPLES)` short-read RNA-seq samples originating from healthy adult and fetal retina, cornea, and RPE to generate the first pan-eye transcriptome annotation. We identify hundreds of novel gene isoforms and examine their significance in the context of ocular development and disease.

# Introduction

|              The transcriptome is defined as the set of unique RNA transcripts expressed in a biological system. A single gene can have multiple distinct transcripts, or isoforms, and there are multiple biological processes that drive the formation of these isoforms including alternative promoter usage, alternative splicing, and alternative polyadenylation. Gene isoforms can have distinct and critical functions in biological processes like development, cell differentiation, and cell migration [-@dykes_hic2_2018], [-@trapnell_transcript_2010], [-@mitra_splicing_2020]. Alternative usage of isoforms has also been implicated in multiple diseases including cancer, cardiovascular disease, Alzheimer’s disease and diabetic retinopathy  [-@vitting-seerup_landscape_2017], [-@neagoe_ciprian_titin_2002], [-@mills_rna-seq_2013], [-@perrin_diabetic_2005].

|              Accurate annotation of gene isoforms is fundamental for understanding their biological impact. For example, while the Gencode human comprehensive transcript annotation (release 28) contains `r toString(NUM_PC_GC_TX)` protein coding and `r toString(NUM_NC_GC_TX)` noncoding transcripts across `r toString(NUM_PC_GC_GENE)` genes and `r toString(NUM_NC_GC_GENE)` pseudogenes, but this annotation is incomplete [-@frankish_gencode_2019], [-@zhang_incomplete_2020]. Some of the first high throughput methods to find novel gene isoforms used short-read (~100bp) RNA-seq to identify novel exon-exon junctions and novel exon boundaries based soley on RNA-seq coverage [-@nagalakshmi_transcriptional_2008]. More recently, several groups have developed specialized tools to use RNA-seq to reconstruct the whole transcriptome of a biological sample, dubbed *de novo* transcriptome construction [-@haas_novo_2013],[-@trapnell_transcript_2010], [-@pertea_stringtie_2015].

|              *De novo* transcriptome construction uses short-read RNA-seq to reconstruct full-length mRNA transcripts. However, a large number of samples are necessary to overcome the noise and short-read lengths of this type of data. Because of increasingly inexpensive sequencing cost, datasets of the necessary size are now available. For example, one of the most comprehensive *de novo* transcriptome projects to date is CHESS, which uses the GTEx data set to construct *de novo* transcriptomes in over 9000 RNA-seq samples from `r toString(NUM_BODY_TISSUES)` distinct body locations to create a comprehensive annotation of mRNA transcripts across the human body [-@gtex_consortium_genetic_2017], [-@pertea_chess_2018]. However, since the GTEx dataset does not include samples from any ocular tissues, the CHESS database remains an incomplete annotation of the human transcriptome.

|              Despite the increasing number of tools developed, there is no gold standard to evaluate the precision and sensitivity of *de novo* transcriptome construction on real (not simulated) biological data. Long-read sequencing technologies provide a potential solution to this problem as long-read sequencing can capture full length transcripts and thus, can be used to identify a more comprehensive range of gene isoforms. While previous iterations of long-read sequencing technologies typically had higher error rates, the new PacBio Sequel II system sequences long-reads as accurately as short-read based sequencing [-@wenger_accurate_2019].

|              We propose that long-read based transcriptomes can serve as a ground truth for evaluating short-read based transcriptomes. In this study, we used PacBio long-read RNA sequencing to inform the construction of short-read transcriptomes. We generated PacBio long-read RNA-seq along with matched Illumina short-read RNA-seq data from an induced pluripotent stem cell (iPSC)-differentiated retinal pigmented epithelium (RPE) cell line. We then designed a rigorous StringTie-based pipeline that maximizes the concordance between short and long-read *de novo* transcriptomes.

|              Finally, we applied this optimized pipeline to a data set containing `r toString(NUM_EYE_SAMPLES)` ocular tissue samples compiled from mining previously published, publicly available short-read RNA-seq data [-@swamy_eye_2019]. We built transcriptomes for three major ocular tissues: cornea, retina, and RPE, using RNA-seq data from both adult and fetal tissues to create a high-quality pan-eye transcriptome. In addition to ocular samples, we used a subset of the GTEx data set to construct transcriptomes for tissues in `r toString(NUM_BODY_TISSUES)` other locations across the body.

|              We used our gold-standard informed pan-eye *de novo* transcriptome to reveal hundreds of novel gene isoforms in the eye and analyze their potential impact on ocular biology and disease. We provide transcript annotation derived from our *de novo* transcriptomes as a resource to other researchers through an R package.

# Results

![](../clean_data/plots/DNTX_flowchar_rework_final.png)

::: {custom-style="CustomCaption"}
Figure 1. Workflow for long-read informed *de novo* transcriptome construction and analysis.x
:::

## Long-read PacBio RNA sequencing guides short-read *de novo* transcriptome construction

|              To evaluate the accuracy of short-read transcriptome construction, we first generated PacBio long-read RNA-seq data and Illumina short-read RNA-seq data from iPSC-RPE (Fig 1). These cells were differentiated using an optimized protocol, and thus minimal biological variation is expected  [-@blenkinsop_human_2015], [-@maruotti_small-moleculedirected_2015]. We used these sequencing data to construct a long-read transcriptome and a short-read transcriptome. In our long-read transcriptome we found `r toString(NUM_TOTAL_LR_TX)` distinct transcripts, and in our short-read transcriptome `r toString(NUM_TOTAL_SR_TX)` distinct transcripts

```{r longread_results_prep}
plot_violin <- function(ldf, origin){
 ldf <- ldf %>% filter(`Transcript Origin` == origin)
 sum_counts <- ldf%>% 
  mutate(intersection_case = gsub('stringtie', 'StringTie', intersection_case),
      intersection_case = gsub('pacbio', 'PacBio', intersection_case),
      intersection_case = gsub('StringTie-PacBio', 'PacBio &\nStringTie', intersection_case)) %>% group_by(intersection_case, `Transcript Origin`) %>% count()
 violin_plot <- ldf %>% 
  mutate(intersection_case = gsub('stringtie', 'StringTie', intersection_case),
      intersection_case = gsub('pacbio', 'PacBio', intersection_case),
      intersection_case = gsub('StringTie-PacBio', 'PacBio &\nStringTie', intersection_case)) %>% 
  ggplot() +
  scale_y_continuous(breaks = c(2000,4000,6000, 8000, 10000)) +
  geom_text_repel(data = sum_counts, aes(x=intersection_case, y = 11000, label=n ), stat= 'identity', nudge_y = -.5) +
  geom_violin(aes(x=intersection_case,y=length)) + geom_hline(yintercept = 2000, color='red') +
  cowplot::theme_cowplot() + ylab(glue('{origin}\nTranscript Length (bp)')) + xlab('') + coord_flip(ylim = c(0,11500)) 
 return(violin_plot)
}
violin_plot_novel <- plot_violin( length_df_plotting, 'Novel')
violin_plot_gencode <- plot_violin(length_df_plotting, 'Gencode')
tpm_exp_plot <- ggplot(data = tpm_df_plotting %>% ungroup, aes(x=co, y=acc, color=`Length Interval`)) + 
 geom_point() + 
 geom_line()+
 #xlim(c(-3,26))+
 #ylim(c(0,1.3))+
 #geom_label_repel(data = labdf_plotting%>% ungroup, aes(x=co, y=y, label = total_tx), fill='white', color='black', nudge_x = -1.5, nudge_y = .25) +
 labs(label = 'asf')+
 ylab('Build Accuracy\n(StringTie ∩ PacBio / StringTie)') +
 xlab('TPM threshold') +
 facet_wrap(~filtering_type) +
 cowplot::theme_cowplot() +
 scale_color_viridis_d()



```
```{r longread_results, fig.height=7, fig.width=8, include=T, echo = F}
violin_plot_novel/violin_plot_gencode/tpm_exp_plot +plot_annotation(tag_levels = 'A') + plot_layout(heights = c(1,1,1.7))
```

::: {custom-style="CustomCaption"}
Figure 2. Transcript length and expression dictate transcriptome construction accuracy. A,B) Distributions of novel(A) and previously annotated(B) transcript lengths between PacBio (long-read) and Stringtie (short-read) transcriptomes. Each distribution is labeled with the total number of transcripts in the distribution  C) short-read construction accuracy stratified by transcript length at different Transcripts Per Million (TPM)-based transcript exclusion thresholds. The "merge" method follows the protocol for constructing transcriptomes outlined by the StringTie authors and keeps any transcripts expressed above a specific TPM threshold in at least one samples. The "mean" method used by our pipeline keeps transcripts whose average expression across all samples is above a specific TPM threshold.
:::

|              In our initial comparison between short and long-read transcriptomes, we noticed a low transcriptome construction accuracy (see Methods) of `r toString(NUM_BASE_TXOME_CONST_ACC)`. When we examined the transcript lengths of each build we saw that the two methods show very different transcript length distributions for both novel and previously annotated transcripts, with the short-read build was comprised mostly of smaller transcripts (Fig 2A). As the PacBio data was generated using two different libraries for 2000 bp and >3000 bp transcripts, we expected an enrichment for longer transcripts in the PacBio data set (Supplemental Figure 2). To assess accuracy relative to transcript length, we grouped transcripts by length in 1000 bp intervals, and compared accuracy between each group. We found that accuracy significantly improves for transcripts longer than 2000 bp. The construction accuracy is `r toString(NUM_TXOME_CONST_ACC_2000_ABOVE)` and `r toString(NUM_TXOME_CONST_ACC_2000_BELOW)` for transcripts above and below 2000 bp, respectively.

|              We experimented with various methods to remove spurious transcripts and improve construction accuracy. We first removed transcripts that were expressed <1 TPM in at least one sample as outlined in StringTie's recommended protocol [-@pertea_transcript-level_2016]. This improved construction accuracy to `r toString(NUM_TXOME_MERGE_ACC_2000_ABOVE)` for transcripts longer than 2000bp and `r toString(NUM_TXOME_MERGE_ACC_2000_BELOW)` for transcripts shorter than 2000bp. As this accuracy was still fairly low, we tried different filtering schemes, including experimenting with machine learning-based strategies to identify transcripts that were computational artifacts (data not shown), but we found that the simplest approach with high performance was to retain transcripts that had an average TPM above a specific threshold(Fig 2C). In our downstream pipeline we keep transcripts that have at least an average of 1 TPM across all samples of the same subtissue type as this threshold achieved a build accuracy of `r toString(NUM_FILT_TXOME_ACC_2000_ABOVE)` for transcripts longer than 2000Bp and retained `r toString(NUM_TXOME_FILTER_SIZE)` transcripts within this short-read RPE dataset. 

## Thousands of novel gene isoforms are detected in human subtissue-specific transcriptomes

```{r overall_stats_table_prep}
df <- sample_table %>% 
 left_join(core_tight) %>% 
 filter(!body_location %in% c('Body', 'Brain', 'ESC_Stem.Cell.Line', 'Lens_Stem.Cell.Line')) %>%
 group_by(subtissue) %>%
 summarise(`Samples` = n(), `Studies` = length(unique(study))) %>% 
 arrange(desc(`Studies`)) %>% 
 inner_join(tx_counts %>% select(subtissue, `Transcriptome Count`=filtered)) %>% 
 mutate(subtissue=gsub('_|\\.',' ', subtissue)) %>% 
 mutate(subtissue = gsub(' Tissue', '', subtissue)) %>% 
 separate(subtissue, into = c('Tissue' , 'Source'), sep = ' ')

```
```{r overall_stats_table, include = T}
df %>% flextable %>% width(5, 2) 
```

::: {custom-style="CustomCaption"}
Table 1. Ocular sample dataset overview and transcriptome count. Transcriptome count is defined as the number of unique transcripts expressed in a given tissue type
:::

|              We built transcriptomes from `r toString(NUM_EYE_SAMPLES)` published, publicly available ocular tissue RNA-seq samples curated in EiaD using an efficient Snakemake pipeline [-@koster_snakemakescalable_2012]. We included both adult and fetal samples from cornea, retina, and RPE tissues mined from `r toString(NUM_STUDIES)` different studies (Table 1). Our fetal tissues consist of both human fetal tissues and human iPSC-derived tissue, as stem cell-derived tissue has been showed to closely resemble fetal tissue [-@klimanskaya_derivation_2004]. To more accurately determine the tissue specificity of novel ocular transcripts, we supplemented our publicly collated normal (non-disease, non-perturbed) ocular data set with `r toString(NUM_BODY_SAMPLES)` samples from `r toString(NUM_BODY_TISSUES)` body locations across `r toString(NUM_MAJOR_BODY_TISSUES)` major tissues from the GTEx project and constructed transcriptomes for each of these body locations [-@gtex_consortium_genetic_2017]. We refer to each distinct body location as a subtissue here after.

|              After initial construction of transcriptomes, we found `r toString(NUM_REF_TX_BASE_TXOME)` previously annotated transcripts and `r toString(NUM_NOVEL_TX_BASE_TXOME)` novel transcripts detected in at least one of our `r toString(NUM_EYE_SAMPLES+NUM_BODY_SAMPLES)` samples. We define novel as any region of the human genome that has not been previously annotated within the Gencode, Ensembl, UCSC, and Refseq annotation databases [-@frankish_gencode_2019] , [-@zerbino_ensembl_2018] , [-@oleary_reference_2016]. After using the filtering methods described above, we merged all subtissue specific transcriptomes into a single final transcriptome which contains `r toString(NUM_TOTAL_TX_BODY)` distinct transcripts with `r toString(NUM_REF_TX_BODY)` previously annotated and `r toString(NUM_NOVEL_TX_BODY)` novel transcripts, and includes `r toString({NUM_TOTAL_NOVEL_SEQ/1000000} %>% round(digits=1))` megabases of previously unannotated genomic sequence (Table 1). We refer to the final pan-body transcriptome as the DNTX annotation hereafter.

|              We split novel transcripts into two categories: novel isoforms, which are novel variations of known genes, and novel loci, which are previously unreported, entirely novel regions of transcribed sequence (Fig 3B). Novel isoforms are further classified by the novelty of their encoded protein: isoforms with novel open reading frame, novel isoforms with a known ORF, and isoforms with no ORF as noncoding isoforms (Fig 3A). The number of distinct ORFs was significantly less than the number of transcripts, with `r toString(NUM_REF_ORF)` previously annotated ORFs and `r toString(NUM_NOVEL_ORF)` novel ORFs across all subtissues. Furthermore, across all subtissues there was an average of `r toString(NUM_AVG_NOVEL_ISOFORM_PER_TISSUE %>% round(0))` novel isoforms and `r toString(NUM_AVG_NOVEL_ORF_PER_TISSUE%>% round(0))` novel ORFs.

```{r overall_stats_figure_prep}

novel_isoforms_per_tissue <- inner_join(novel_transcripts_per_tissue, novel_orfs_per_tissue) %>% 
 select(-novel_pc_count) %>% 
 gather(key = 'transcript_type', value = 'counts', -body_location, -color) %>% 
 filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
 mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location_pretty=gsub('Tissue','', body_location_pretty),
     body_location_pretty=gsub('avg','mean', body_location_pretty),
     body_location_pretty=gsub('\\(',' (', body_location_pretty),
     body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
     body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
 mutate(transcript_type = gsub('_',' ', transcript_type),
     transcript_type = gsub(' count', '', transcript_type),
     transcript_type = gsub('nc','NC', transcript_type),
     transcript_type = gsub('orf','ORF', transcript_type),
     transcript_type = gsub('novel','Novel', transcript_type),
     transcript_type = gsub('ref','Ref.', transcript_type))
novel_loci_per_tissue <- novel_loci_per_tissue %>% 
 gather(key = 'transcript_type', value = 'counts', -body_location, -color) %>% 
 filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
 mutate(transcript_type = gsub('_\\w+','', transcript_type),
     transcript_type = gsub('noncoding' ,'Noncoding', transcript_type),
     transcript_type = gsub('pc','Protein\nCoding', transcript_type)) %>% 
 mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location_pretty=gsub('avg','mean', body_location_pretty),
     body_location_pretty=gsub('Tissue','', body_location_pretty),
     body_location_pretty=gsub('\\(',' (', body_location_pretty),
     body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
     body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty))
color_list<- novel_isoforms_per_tissue$color
names(color_list) <- novel_isoforms_per_tissue$body_location_pretty

loci <- ggplot(data = novel_loci_per_tissue) +
 geom_col(aes(x=body_location_pretty, y=counts, fill = body_location_pretty, alpha = transcript_type), position = 'dodge') +
 scale_fill_manual(values = color_list, 'Body\nLocation') +
 scale_alpha_discrete(range=c(.5,1), name = 'Transcript\nType') +
 #ggtitle('Novel Isoforms Contructed Across the Body')+
 ylab('Novel Loci\nCount')+
 xlab('Body Location')+
 cowplot::theme_cowplot() + 
 theme(axis.text.x=element_blank(), panel.background = element_blank())

isotype_per_tissue <- ggplot(data = novel_isoforms_per_tissue) +
 geom_col(aes(x=transcript_type, y=counts, fill = body_location_pretty), position = 'dodge') +
 scale_fill_manual(values = color_list) +
 scale_alpha_discrete(range=c(.5,1)) +
 #ggtitle('Novel Isoforms Contructed Across the Body')+
 ylab('Number of Novel Isoforms')+
 xlab('')+
 cowplot::theme_cowplot() + theme(legend.position = "none") + 
 theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank())+
 facet_wrap(~body_location_pretty, ncol=2)

novel_transcript_types <- novel_isoform_anno_df %>% filter(label != 'Protein Coding') %>% 
 mutate(group = 'Transcript Type') %>% 
 rename(`Transcript Type` = label) %>% 
 ggplot(data = ., aes(x = group, fill = `Transcript Type`, y=count)) +
 geom_bar(position = 'fill',stat = 'identity') +
 geom_text(aes(label = count), position = position_fill(vjust=.5)) +
 ylab('% of all novel transcripts') +
 cowplot::theme_cowplot() + 
 theme(panel.background = element_blank(), axis.title.x = element_blank())

novel_exon_types <- exon_type_by_transcript_type %>% group_by(nv_type_rc) %>% 
 summarise(count = n()) %>% 
 mutate(group = 'Novel Exon Type') %>% 
 rename(`Novel Exon Type` = nv_type_rc) %>% 
 ggplot(data = ., aes(x = group, fill = `Novel Exon Type`, y=count)) +
 geom_bar(position = 'fill',stat = 'identity') +
 geom_text(aes(label = count), position = position_fill(vjust=.5)) +
 ylab('% of all novel transcripts') +
 cowplot::theme_cowplot() +
 theme(panel.background = element_blank(), axis.title.x = element_blank())

exon_type_tx_plot <- exon_type_by_transcript_type %>% 
 mutate(comp_transcript_type = str_replace(comp_transcript_type, 
                      'Novel Isoform Novel Protein', 
                      'Novel Isoform Novel ORF')) %>% 
 mutate(comp_transcript_type = gsub('Isoform', 'Isoform,\n', comp_transcript_type),) %>% 
 mutate(nv_type_rc = gsub('_',' ', nv_type_rc),
     nv_type_rc = gsub('novel', 'Novel', nv_type_rc), 
     nv_type_rc = str_replace_all(nv_type_rc, 'TES', 'TE'),
     nv_type_rc = str_replace_all(nv_type_rc, 'TSS', 'FE')) %>% 
 group_by(comp_transcript_type, nv_type_rc) %>% 
 summarise(count = n()) %>% 
 ggplot(data=., aes(x = comp_transcript_type,y=count, fill = nv_type_rc)) +
 geom_bar(position = 'fill', stat = 'identity')+
 geom_text(aes(label = count), position = position_fill(vjust=.5)) +
 ylab('Proportion of\nNovel Isoforms') +
 cowplot::theme_cowplot() +
 theme(axis.text.x=element_text(angle=45, hjust = 1), 
    panel.background = element_blank(), axis.title.x = element_blank()) +
 scale_fill_manual(values = pals::cols25() %>% unname(), name = 'Transcript\nCategory')

```
```{r overall_stats_figure, fig.height=9, fig.width=8, include= T, echo = F}
layout <- '
AB
AC
'
isotype_per_tissue + loci + exon_type_tx_plot +plot_annotation(tag_levels = 'A') +plot_layout(design = layout, heights = c(1,5), widths = c(1.5,1), guides = 'collect') 
```

::: {custom-style="CustomCaption"}
Figure 3. Overview of novel isoforms. A) Number of novel gene isoforms, grouped by transcript type. Brain and body represent an average of 13 and 34 distinct subtissues, respectively. B) Novel protein coding and noncoding loci. Novel exon composition of novel isoforms, by isoform type. Labels indicate number of transcripts. C) Classification of novel exon types, stratified by novel isoform type.
:::

|              Novel isoforms can occur due to an omission of a previously annotated exon, commonly referred as exon skipping or the addition of an unannotated exon which we refer to as a novel exon. We further classified novel exons by the biological process that may be driving their formation: alternative promoter usage driving the addition of novel first exons (FE), alternative polyadenylation driving the addition of novel terminal exons (TE), and alternative splicing driving the formation of all novel exons that are not the first or last exon  [-@landry_complex_2003], [-@tian_alternative_2017], [-@wang_mechanism_2015]. We then split alternatively spliced exons into their commonly seen patterns, alternative 5' splice site (A5SS), alternative 3' splice site (A3SS), and retained introns (RI). Exons whose entire sequence was unannotated and is not a retained intron are fully novel exons. We note that all three of these mechanisms can lead to exon skipping, so for simplicity we grouped all novel isoforms resulting from exon skipping together. We found that the majority of novel exons within our dataset are novel FEs. We noticed that the majority of RI exons lead to novel ORFs, whereas novel isoforms with omitted exons more often lead to noncoding isoforms. (Fig 3C)

## *De novo* transcriptomes match previously published experimental data better than existing annotation

|              We validated *de novo* transcriptomes using three independent approaches. We first looked for evolutionary conservation since it is commonly accepted as a proxy for functional significance.  We used the PhyloP 20 way species alignment, a measure of conservation between species, to calculate the average conservation score for each exon in the DNTX annotation and compared that to the average conservations score for each exon in the GENCODE annotation [-@pollard_detection_2010]. We found that, on average, exons in the DNTX annotation are more conserved than exons in the GENCODE annotation (pvalue <2.2e-16) (Supplemental Figure 2A). 

|              Next, since we observed an enrichment in novel first and last exons within our data set, we decided to compare the TSS and TES within the DNTX annotation to two well-established annotation databases from FANTOM and the polyA Atlas [-@noguchi_fantom5_2017], [-@herrmann_polyasite_2020]. We compared DNTX and GENCODE TSS’s to CAGE-seq data from the FANTOM consortium; as CAGE-seq is optimized to detect the 5’ end of transcripts, we reasoned that it can serve as a valid ground truth set to evaluate TSS detection [-@takahashi_cage-_2012]. We calculated the absolute distance of DNTX TSS’s to CAGE peaks, and compared them to the absolute distance of GENCODE TSS’s to CAGE peaks. We found that, on average, DNTX TSS’s were closer to CAGE peaks than GENCODE TSS’s (pvalue <2.2e-16)(Supplemental Figure 2B). 

|              Finally, we evaluated TES’s using the polyA Atlas, which is comprised of polyadenylation signal annotation generated from aggregating 3’ seq data from multiple studies. As 3’-seq data is designed to accurately capture the 3’ ends of transcripts, it can similarly serve as a ground truth set to evaluate the accuracy of TES’s [-@beck_3-end_2010]. We calculated the absolute distance of DNTX TES's to annotated polyA signals and compared them to the absolute distance of GENCODE TES's to polyA signals. We found that on average DNTX TES's are closer to annotated polyadenylation signals than gencode TSS's (pvalue <2.2e-16) (Supplemental Figure 2C)

## *De novo* transcriptomes reduce overall transcriptome sizes

|              *De novo* transcriptomes removed on average `r toString(NUM_AVG_TXOME_SIZE_DECREASE*100)` % of a subtissue’s base transcriptome. We defined base transcriptome for a subtissue as any transcript in the GENCODE annotation with non-zero TPM in at least one sample of a given subtissue. This was a large reduction in transcriptome size and we wanted to ensure that we were not unduly discarding data. We quantified transcript expression of each sample using Salmon with two methods: once using the full gencode v28 human transcript annotation, and once using its associated subtissue specific transcriptome. We found that despite the `r toString(NUM_AVG_TXOME_SIZE_DECREASE*100)` % reduction in number of transcripts between the base gencode and *de novo* transcriptomes (Supplemental Figure 3A), the average Salmon mapping rate increased by `r toString(NUM_AVG_MAP_RATE_DIF*100)` % indicating that the vast majority of gene expression data is retained within our transcriptome (Supplemental Figure 3B).

## Novel Isoforms are identified in ocular tissues

```{r, novel_isoforms_ocular_tissues_prep}
source('scripts/R_raincloud.R')
plot_list <- novel_eye_tx_by_tissue[!names(novel_eye_tx_by_tissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line') ]
all_tx <- tibble(transcript_id = reduce(plot_list, union))
plot_df <- bind_cols(all_tx, lapply(plot_list, function(x) all_tx$transcript_id%in% x))
FIG_FONT_SIZE=12

colnames(plot_df) <- gsub('_|\\.', ' ', colnames(plot_df) )               
colnames(plot_df) <- gsub('Adult', '(Adult)', colnames(plot_df) )
colnames(plot_df) <- gsub('Fetal', '(Fetal)', colnames(plot_df) )
colnames(plot_df) <- gsub('Tissue', '', colnames(plot_df) )

us <- ComplexUpset::upset(data = plot_df, 
             intersect = colnames(plot_df)[-1], 
             min_size= 900, 
             set_sizes = F,
             #height_ratio = .8, 
             #width_ratio = .1,
             wrap = T,
             stripes = c('white','white'),
             base_annotations = list(
              'Intersection size' = intersection_size(text_colors = c(on_background='red', on_bar='red'))
             ),
             themes=upset_modify_themes(
              list(
               "intersections_matrix"= theme(text = element_text(size=FIG_FONT_SIZE) ),
               "Intersection size"= theme(text = element_text(size=FIG_FONT_SIZE),
                             panel.grid.major = element_blank(), 
                             panel.grid.minor = element_blank()),
               "overall_sizes"= theme(axis.text.x=element_text(angle=45),text = element_text(size=FIG_FONT_SIZE) ),
               "default"= theme(text = element_text(size=FIG_FONT_SIZE)) 
              )
             )
             
)

plotting_piu_df <- piu_df %>% mutate(new_stage=ifelse(stage == 'adult', 'Adult', 'Fetal'), 
                   subtissue=paste0(tissue, '_', new_stage, '.Tissue')) %>% 
 mutate(subtissue=gsub('_|\\.', ' ', subtissue),
     subtissue=gsub('avg','mean', subtissue),
     subtissue=gsub('Tissue','', subtissue),
     subtissue=gsub('\\(',' (', subtissue),
     subtissue=gsub('Adult','(Adult)', subtissue),
     subtissue=gsub('Fetal','(Fetal)', subtissue))

piu <- ggplot(data=plotting_piu_df, aes(x=stage, y=piu, fill=subtissue, color=subtissue)) + 
 geom_flat_violin(position = position_nudge(x = .2, y = 0),adjust = 2)+
 geom_point(position = position_jitter(width = .15), size = .25, alpha=.2) + 
 geom_boxplot(outlier.shape = NA, alpha = 1, width = .1, colour = "BLACK")+ 
 scale_fill_manual(values = color_list) +
 scale_color_manual(values = color_list) +
 ylab('FIU')+xlab('tissue')+coord_flip()+ 
 cowplot::theme_cowplot()+ facet_wrap(~tissue, nrow = 3) +
 theme(legend.position = "none")

```
```{r novel_isoforms_ocular_tissues, fig.width=10, fig.height=5, include=T}
us + piu + plot_layout(widths = c(2,1)) + plot_annotation(tag_levels = 'A') & theme(plot.tag = element_text(size= FIG_FONT_SIZE), text = element_text(size = FIG_FONT_SIZE))

```

::: {custom-style="CustomCaption"}
Figure 4. Overview of novel gene isoforms in the eye. A) Set intersection of novel isoforms in ocular transcriptomes. B) Boxplots of fraction isoform usage (FIU) overlaid over FIU data points with estimated distribution of data set above each boxplot.
:::

|              Using the pan-eye transcriptome, we compared the overlap in constructed novel isoforms across ocular subtissues and found that `r toString(NUM_TSPEC_ISO)` % of novel isoforms are specific to a singular ocular subtissue (Fig 4A). Additionally, fetal-like tissues had more novel isoforms that their adult counterpart. For each novel isoform we then calculated fraction isoform usage (FIU), or the fraction of total gene expression a transcript contributes to its parent gene. We found that, on average, novel isoforms contributed to `r toString(NUM_AVG_PIU)` % of their parent gene's expression but in each subtissue we found multiple novel isoforms that contribute to most of their parent genes expression (Fig 4B)

## Differential usage of gene isoforms occurs during retinal development

|              Multiple studies have shown that gene isoforms play a significant role in eye development [-@bharti_alternative_2008], [-@mellough_integrated_2019]. We hypothesized that the DNTX annotation provides additional insight into alternative isoform usage and identifies novel gene isoforms potentially involved in eye development. We used RNA-seq data of the developing retina from Mellough et al, an independent data set that we did not include for transcriptome construction, and used a subset of the DNTX annotation corresponding to fetal retina to quantify transcript expression and identify transcripts with significant changes in expression across retinal development. Transcripts that are differentially expressed (qvalue <.01) and have a mean FIU difference of .25 in at least one comparison of time points are indicative of differential transcript usage (DTU).

```{r fetal_retina_diff_exp, fig.width=13, fig.height=15, include=T, echo=F}

ddf <- tibble(x=c(0,0), y=c(0,0), sig = c('DTU', 'NOT DTU'))
vp <- vp + 
    geom_point(data = ddf, aes(x=x, y=y, color = sig), alpha = 0) + 
    scale_color_manual(values = c( 'DTU'= 'red', 'NOT DTU'='lightgrey')) + 
    guides(color = guide_legend(title = NULL, override.aes = list(alpha = 1, size = 3)))
bdes <- '
AB
AC
AC
AC
AC
AC
AC'
bottom <- gghm +gm_fetret + piu_fetret +plot_layout(design = bdes)

top <- vp |dp
full_des <-'
AAA
BBB
BBB'
fp <- top/bottom + 
  plot_layout(design = full_des) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size=13, face = 'bold'), legend.margin = margin(0,0,0,0), legend.box.margin=margin(-10,-10,-10,-10))
fp

```

::: {custom-style="CustomCaption"}
Figure 5 Differential Transcript usage during retinal development. A) Volcano plot of tested transcripts B) Dot plot for gene set enrichment analysis C) Heatmap of hiearchical clustering of transcripts with DTU associated with eye development D) Transcript models for *MYO9A*, a gene undergoing DTU E) FIU change in *MYO9A* FIU across development F) average log-transformed TPM expression of *MYO9A* across retinal development 
:::

|              We analyzed `r toString(NUM_DEVRET_SAMPLES)` samples across `r toString(NUM_DEVRET_TIMESPOINTS)` developmental days post fertilization and found `r toString(NUM_DTU_TX)` transcripts across `r toString(NUM_DTU_GENE)` genes displaying DTU (Fig 5A). We found that genes involved in DTU are enriched(qvalue <.05) for genes related to eye and neurological development (Fig 5B), and that hierarchical clustering of DTU transcripts generates an early stage and late stage cluster (Fig 5C). One of these genes, *MYO9A*, is a classical example of DTU. *MYO9A* is associated with the visual perception GO term, plays a role in ocular development, and has been associated with ocular disease [-@gorman_cloning_1999]. While expression of *MYO9A* remains relatively unchanged across development, expression of two of its associated isoforms in fetal retina (Fig 5D) changes dramatically during development: a novel isoform is highly expressed early during development, but switched to the canonical isoform later in development (Fig 5E,F). This novel isoform contains a novel exon within the protein coding region of the isoform as well as novel last exon extending the 3’ UTR (Fig 5d). A full list of genes and transcripts displaying  DTU is available in Supplemental Data (Supplemental Data 4).

## *De novo* transcriptomes allow for a more precise variant prioritization.

|              The identification of a disease-causing variant through genome sequencing is a common step in diagnosing genetic disease, when disease causing variants cannot be determined from exonic sequencing. Prediction of a variant's biological impact and subsequent variant prioritization is a fundamental step in this process. Many methods for predicting variant effects on protein function or gene expression are based on location within the body of a transcript; for example variants that disrupt splice sites and start/stop codons are considered to be the most damaging, while variants within intronic and intergenic regions have unknown impact or are not classified, and, thus, are not included for further consideration. However, multiple studies have identified pathogenic deep intronic variants for retinal dystrophies  [-@braun_non-exomic_2013], [-@bauwens_abca4-associated_2019], [-@zernant_analysis_2014], [-@sangermano_deep-intronic_2019], [-@jamshidi_contribution_2019], [-@mayer_homozygosity_2016], [-@geoffroy_whole-genome_2018]. Pathogenic intronic variants are thought to function by introducing a novel splice site, disrupting regulatory motifs, or altering a tissue-specific transcript. To explore this third possibility, we mapped known pathogenic intronic variants onto novel isoforms within the *de novo* transcriptomes.

```{r, include = T}
clean_variant_results %>% 
  filter(`Location(hg19)`!="Chr1:94468019", !duplicated(`Location(hg19)`)) %>% 
  mutate(`Canonical Variant HGVS` = paste(variant, NM_accession, sep = ', ')) %>% 
  select(`Gene Name`, `Associated Disease`,`Location (hg19)`=loc_with_var,
      `Canonical Variant HGVS`,
      `Gencode Predicted Consequence`, `DNTX Predicted Consequence`, `Published Study` ) %>% 
  arrange(`Gene Name`, `Associated Disease`) %>% 
  flextable() %>% 
  align(align='left', part = 'all') %>% 
  theme_vanilla %>% 
  width(1, .65) %>% 
  width(2, 1.0) %>%
  width(3, 1.0) %>% 
  width(4, 1.0) %>% 
  width(5, 1.4) %>% 
  width(6, 1.4) %>% 
  width(7, 1.1) %>% 
  merge_v(j = c("Gene Name", "Associated Disease", 'Published Study')) %>%
  fontsize(part = 'body', size = 8)
```

::: {custom-style="CustomCaption"}
Table 2. Pathogenic variants previously considered intronic that are on expressed transcripts in the retina *de novo* transcriptome. Canonical  human genome variation society (HGVS) annotation is based on transcripts from the RefSeq annnotation. Predicted consequences were generaed with the Variant Effect Predictor(VEP)
:::

|              We used a list of 129 intronic and noncoding variants previously identified as pathogenic for a retinal dystrophy and predicted the effect of these variants with Ensembl’s Variant Effect Predictor using a subset of the DNTX annotation corresponding to fetal and adult retina as the input transcript annotation. We identified ten variants whose predicted effect increased in severity due the presence of a novel gene isoform in a previously intronic region (Table 2). Seven of these variants were in deep intronic hotpsots known for pathogenic variation within the gene ABCA4. 

```{r}
load('clean_data/rdata/ABCA4_tx_plot_data.Rdata')
plot_data <- all_plot_df%>% filter(type == 'exon', transcript_id != "DNTX_00008583") %>% 
  mutate(transcript_id = factor(transcript_id, levels = keep_tx[keep_tx!= "DNTX_00008583"  ]))
color_list <- c('black', 'red')
names(color_list) <- c('ref', 'novel')
        
point_type = ifelse(plot_data$strand[1] == '+',62, 60)

k <- plot_data %>% filter(type == 'exon')
tx_plot <- ggplot(data = plot_data) +
    geom_hline(yintercept = 0, colour='black', size=2)+
    geom_rect(aes(xmin=Xmin, xmax=Xmax, ymin=Ymin, ymax=Ymax,fill=`exon type`))+
    #geom_rect( aes(xmin=Xmin, xmax=Xmax, ymin=Ymin, ymax=Ymax,fill=`exon type`))+
    xlim(c(min(k$Xmin), max(k$Xmax))) +
    scale_fill_manual(values = color_list) +
    facet_wrap(~transcript_id, ncol=1, switch = 'y')+
    theme_void() +#+
    theme(strip.text.y.left = element_text(angle = 0))
    #theme(strip.text.y = element_text(angle = 180))


var_plot_df<- all_plot_df %>% filter(type == 'variant') %>% select(var_id=novel_exon_id, Xmin, Xmax) %>% 
    inner_join(clean_var_abca4_only) %>% 
    filter(!duplicated(var_id)) %>% 
    mutate(Y=1.5, hg19_l = str_split(raw_loc, ':') %>% sapply(function(x) x[2] ), 
           gene= gene_name, hg38_l =str_split(Location, ':') %>% sapply(function(x) x[2] ) )


k <- plot_data %>% filter(type == 'exon')
j <- max(k$Xmax)+3.5
vars <- ggplot(var_plot_df)+
    geom_hline(yintercept = 0, colour='black', size=2)+
    geom_point(aes(x=Xmin, y=.1), shape = 8 )+
    geom_label_repel(data = var_plot_df %>% filter(!duplicated(hg19_l)), aes(x=Xmin, y=.1, label=hg19_l), nudge_y = 1.5)+
    xlim(c(min(k$Xmin), max(k$Xmax))) +
    ylim(c(0,.4)) +
    geom_segment( x=j, xend = j , y=0, yend=.15)+
    geom_segment(x=j - (j-min(k$Xmin))/10,  xend = j, y=.15, yend=.15  )+
    geom_point(x =j - (j-min(k$Xmin))/10, y=.15, shape=60, size=5 )+
    facet_wrap(~gene, switch = 'y')+
    theme_void() +
    theme(strip.text.y.left = element_text(angle = 0, size = 20))
des= '
AA
BB
BB
BB
'

```


```{r, include = T, fig.height=4, fig.width=12}
vars/tx_plot+plot_layout(design = des)
```

::: {custom-style="CustomCaption"}
Figure 6. Transcript models for selected Isoforms of *ABCA4* along with location of pathogenic intronic variants. Location is on the hg19 human genome build. Thick lines indicate protein coding regions. Arrow indicates direction of transcription. Introns not drawn to scale
:::

|              These variants were spanned by three distinct novel isoforms with two containing open reading frames (ORFs) encoding only the carboxy-terminus of the canonical protein isoform, and one noncoding spanning the proximal half of the canonical isoform (Fig 6).  *ABCA4* expression and function has also been observed in RPE [-@lenis_expression_2018]. However, we did not observe these transcripts in RPE, suggesting that these pathogenic variants are primarily affecting retinal-specific *ABCA4* transcripts. We note that these transcripts have not been experimentally validated.

|              To further highlight the potential importance of *de novo* transcriptomes for future genetic tests we determined how many genes associated with retinal disease from RetNet have novel isoforms (sph.uth.edu/retnet/). We found that within the set of genes with novel isoforms, there is significant enrichment of retinal disease genes  (hypergeometric pvalue = `r toString(NUM_RETNET_ENRICH_PVAL)`), with `r toString(NUM_RETNET_IN_DNTX)` out of `r toString(NUM_RET_NET_GENES)` RetNet genes having a novel isoform. A full list of these genes is available in the Supplementary data(supplemental data 5).

## A companion visualization tool enables easy use of *de novo* transcriptomes

```{r app_viz, fig.height=6, include=T, echo =F}
piu <- as_ggplot(rasterGrob(readPNG('clean_data/plots/piu_shiny_app.png'), interpolate = T))
frac_det <- as_ggplot(rasterGrob(readPNG('clean_data/plots/frac_sample_det.png'), interpolate = T))
gene_body <- as_ggplot(rasterGrob(readPNG('clean_data/plots/gene_bodies_shiny_app.png'), interpolate = T))
layout <- '
AABB
CCCC
'
piu + frac_det + gene_body + plot_layout(design = layout)+ plot_annotation(tag_levels = 'A')

```

::: {custom-style="CustomCaption"}
Figure 7. Screenshots from dynamic *de novo* transcriptome visualization tool. A). FIU bar plot for selected gene and subtissue. B). Exon level diagram of transcript body Thicklines represent coding region of transcript. novel exons colored in red. Tooltip contains genomic location and phylop score C) Bargraph of fraction of samples within dataset each transcript was consructed in by tissue.
:::

|              To make our results easily accessible we designed a R-Shiny app for visualizing and accessing our *de novo* transcriptomes. For each subtissue we show the FIU for each transcript associated with a gene (Fig 7A). We show the exon-intron structure of each transcript and mousing over exons show genomic location overlapping SNPs, and phylogenetic conservation score (Fig 7B). We additionally show a barplot of the fraction of samples each transcript was constructed in (Fig 7C). Users can also download the *de novo* transcriptomes for selected subtissues in GTF and fasta format. Instructions to download and run the app are available at https://github.com/vinay-swamy/ocular_transcriptomes_shiny. While visualization of direct transcript expresion is not a part of this app, it can be viewed in the eyeIntegration app [-@swamy_eye_2019] by selected 'DNTX' as the transcript annotation. Finally, we package all tools used for our transcriptome pipeline within a portable docker container with a stand-alone run script. This pipeline allows other researchers to run their own samples, and generate figures and annotations similar to what is shown here, available at https://github.com/vinay-swamy/ocular_transcriptomes_pipeline

# Discussion

|              Motivated by the lack of a comprehensive transcriptome for the eye, we constructed transcriptomes for adult and fetal retina, RPE and cornea. By using long-read RNA-seq data to calibrate our short-read construction pipeline, we were able to identify biologically relevant transcriptomes. We found that concordance between long and short-read-based transcriptomes is directly related to transcript length and transcript expression. We saw a clear inability within the PacBio data set to accurately detect transcripts shorter than 2000bp for both previously annotated and novel transcripts. As many of the transcripts constructed using short-reads are below this threshold, long-read sequencing data enriched for smaller transcript sizes would provide greater insight in future studies.

|              We used a large dataset compiled from published RNA-seq data to build the pan-eye transcriptomes, an approach that has several key advantages. First, the large sample size overcomes the noisy nature of RNA-seq data. Second, as the cohort is constructed from many independent studies, we are more confident that the transcriptomes accurately reflect the biology of their originating subtissue and are not a technical artifact due to preparation of the samples. As another line of evidence, the *de novo* transcriptomes match existing large scale data sets and are more conserved than existing annotations (Supplemental Figure 2).

|              In each ocular subtissue we examined, we found hundreds of novel gene isoforms, many of which were novel due to novel exons. Within ocular subtissues, these novel isoforms are most commonly specific to single subtissue. This makes sense as a majority of the exons in our *de novo* transcriptomes are first and last exons, which have been previously shown to significantly contribute to the tissue specificity of gene isoforms  [-@reyes_alternative_2018]. We also found that on average novel isoforms represent about `r toString(NUM_AVG_PIU)` % of their parent gene's expression. Future studies are needed to identify the function of these isoforms.  One possibility is that some of these isoforms are only expressed in rare cell types, as transcript annotation was previously shown to be incomplete in rare cell types [-@zhang_incomplete_2020]. This especially makes sense in the retina which contain over a dozen distinct cell types, several of which contribute to 5% or less of the total cell population [-@yan_cell_2020]. As we imposed a strict expression filter as part of our transcriptome pipeline, we may have removed transcripts specific to rare cell types.

|              In conclusion, we created the first pan-eye transcriptome annotation and showed that it is useful in understanding the role of gene isoforms in ocular biology and improving the ability to diagnose inherited eye diseases. We hope this work is useful as a starting point for other researchers; [delete] to make the transcriptomes easily accessible to other researchers we designed a webapp both for visualization and to quickly access tissue-specific annotation files. We believe this project will enable other researchers to explore new research directions and answer long pending questions.

## Methods

## Generation of PacBio long-read RNA sequencing data and Illumina short-read RNA sequencing data

|              Human iPSCs were differentiated into RPE using previously described protocols in  [@bryan_identifying_2018] and [@may-simera_primary_2018]. iPSC-derived RPE (iPSC-RPE) cells at 42 days post differentiation were lysed with TRIzol reagent (Thermo Fisher Scientific; cat # 15596026) and total RNA was isolated using the Direct-zol RNA MiniPrep Kit (Zymo Research, Irvine, CA). 5-6 µg total RNA that passed quality control metric (RIN >.9) were used for PacBio library preparation. For PacBio HiFi circular consensus sequencing(CCS), libraries were prepared following the "Procedure-Checklist-Iso-Seq-Express-Template-Preparation-for-Sequel-and-Sequel-II-Systems" protocol. Two libraries were generated: one to capture transcripts 2 kilobases(kb) or smaller, and one to capture transcripts between 2-5kb. Sequencing was done on the PacBio Sequel II system for a movie time of 24 hours. 

|              For Illumina sequencing, Poly-A selected stranded mRNA libraries were constructed from 0.5-1 µg total RNA using the Illumina TruSeq Stranded mRNA Sample Prep Kits according to manufacturer’s instructions. Amplification was performed using 10-12 cycles to minimize the risk of over-amplification.  Unique dual-indexed barcode adapters were applied to each library. Libraries were pooled in equimolar ratio and sequenced together on a HiSeq 4000. At least 57 million 75-base read pairs were generated for each individual library.  Data was processed using illumina Real Time Analysis (RTA) version 2.7.7. All library preparation and sequencing was performed at the National Institutes of Health Intramural Sequencing Center (NISC). 

## Code availability and software versions.

|              To improve reproducibility, all code used for both the analyzing the data and generating the figures for this paper was written as multiple Snakemake pipelines. Each Snakefile contains the exact parameters for all tools and scripts used in each analysis. [-@koster_snakemakescalable_2012] All code (and versions) used for this project is publicly available in the following github repositories: <https://github.com/vinay-swamy/ocular_transcriptomes_pipeline> (main pipeline), <https://github.com/vinay-swamy/ocular_transcriptomes_longread_analysis> (long-read analysis pipeline), <https://github.com/vinay-swamy/ocular_transcriptomes_paper> (figures and tables for this paper), <https://github.com/vinay-swamy/ocular_transcriptomes_shiny> (webapp). Additionally, all Snakefiles are included as supplementary data.(supplementary data files 1-3)

## Analysis of long-read data

|              PacBio sequencing movies were processed into full length, non-chimeric (FLNC) reads using the IsoSeq3 3.1.2 pipeline in the PacBio SMRT link v7.0 software. The existing ENCODE long-read RNA-seq pipeline (https://github.com/ENCODE-DCC/long-read-rna-pipeline) was rewritten as a Snakemake workflow as follows. Transcripts were aligned to the human genome using minimap2(18), using an alignment index built on the gencode v28 primary human genome. Sequencing errors in aligned long-reads were corrected using TranscriptClean (19) with default parameters. Splice junctions for TranscriptClean were obtained using the TranscriptClean accessory script “get_SJs_from_gtf.py” using the gencode v28 comprehensive transcript annotation as the input. A list of common variants to avoid correcting were obtained from the ENCODE portal (https://www.encodeproject.org/files/ENCFF911UGW/). The long-read transcriptome annotation was generated with TALON (20). A TALON database was generated using the talon_initialize_database command, with all default parameters, except for the “–5P” and “–3p” parameters. These parameters represent the maximum distance between close 5’ start and 3’ ends of similar transcript to merge and were both set to 100 to match parameters used in later tools. Annotation in GTF format was generated using the talon_create_GTF command, and transcript abundance values were generated using the talon_abundance command.

## Analysis of short-read RPE data

|              Each sample was aligned to the Gencode release 28 hg38 human genome assembly using the genomic aligner STAR and the resulting BAM files were sorted using samtools sort [-@frankish_gencode_2019],[-@dobin_star_2013],[-@li_sequence_2009]. For each sorted BAM file, a per-sample base transcriptome was constructed using StringTie with the Gencode v28 comprehensive annotation as a guiding annotation [-@frankish_gencode_2019],[-@pertea_stringtie_2015]. All sample transcriptomes were merged with the long-read transcriptome using gffcompare[-@pertea_gff_2020] with default parameters. We note that the default values for the distance to merge similar 5' starts and 3 ends of transcripts in gffcompare is the same to what we chose for TALON. We defined the metric construction accuracy, used to evaluate short-read transcriptome construction as the following: 
$Construction\ Accuracy = \frac {short\ read \ transcriptome\ \cap \ long\ read\ transcriptome} {short\ read \ transcriptome}$

## Construction of subtissue-specific transcriptomes.

|             We constructed transcriptomes for `r toString(NUM_EYE_SAMPLES+NUM_BODY_SAMPLES)` samples in the Eye in a Disk(EiaD), a dataset generated from aggregating publically available healthy, unperturbed RNA-seq samples from `r toString(NUM_BODY_TISSUES + NUM_EYE_TISSUES)` distinct locations of the body across `r toString(NUM_STUDIES)` different studies. Specific information on how this dataset was generated is detailed in the methods from our previous work [-@swamy_eye_2019]. We constructed a transcriptome for each sample, and merged samples together to create `r toString(NUM_BODY_TISSUES + NUM_EYE_TISSUES)` subtissue-specific transcriptomes. We define subtissue as a unique body location and are either temporally different versions of the same tissue(adult vs fetal tissue), or different regions of a larger tissue (cortex vs cerebellum in brain). Tissue refers to complete whole tissue (retina, brain, liver). For each subtissue-specific transcriptome, we removed transcripts that had an average expression less than 1 Transcripts Per Million (TPM) across all samples of the same subtissue type. All subtissue-specific transcriptomes were merged to form a single unified annotation file in general transfer format(GTF) to ensure transcript identifiers were the same across subtissues. We merged all ocular subtissue transcriptomes to generate a separate pan-eye transcriptome.

## Subtissue specific transcriptome quantification

|              For each resulting subtissue specific transcriptome, we extracted transcript sequences using the tool gffread and used these sequences to build a subtissue-specific quantification index using the index mode of the alignment-free quantification tool Salmon [-@pertea_gff_2020], [-@patro_salmon_2017]. For each sample, we quantified transcript expression using the quant mode of Salmon, using a sample's respective subtissue specific quantification index. We similarly quantified all ocular samples using the pan-eye transcriptome and the Gencode v28 reference transcriptome.

## Annotation of novel exons

|              First, a comprehensive set of distinct, annotated exons was generated by merging exon annotation from gencode, ensembl, UCSC, and refseq. We then defined a novel exon as any exon within our transcriptomes that does not exactly match the chromosome, start, end and strand of an annotated exon. Novels exons were classified by splitting exons into 3 categories: first, last, and middle exons. We then extracted all annotated exon start and stop sites from our set of previously annotated exons. Novel middle exons that have an annotated start but an unannotated end were categorized as a novel alternative 3’ end exons and similarly novel middle exons with an unannotated start but annotated end were categorized as a novel alternative 5’ start exons. Novel middle exons whose start and end match annotated exon start and ends were considered retained introns. Novel middle exons whose start and end do not match annotated starts and ends were considered fully novel exons. We then classified novel first and last exons. Novel first exons were first exons whose start is not in the set of annotated exon starts, and novel last exons were terminal exons whose end is not in the set of annotated exon ends. This analysis of novel transcripts is implemented in our Rscript “annotate_and_make_tissue_gtfs.R”.

## Validation of DNTX with phylop, CAGE data, and polyA signals

|              PhyloP scores for the phylop 20-way multi species alignment were downloaded from UCSC's FTP server on October 16th, 2019 and converted from bigWig format to bed format using the wig2bed tool in BEDOPs [-@pollard_detection_2010], [-@neph_bedops_2012]. The average score per exon in both the gencode and DNTX annotation was calculated by intersecting exon locations with phylop scores and then averaging the per base score for each exon, using the intersect and groupby tools from the bedtools suite, respectively. Significant difference in mean phylop score was tested with a Mann Whitney U test. 

|              CAGE peaks were download from the FANTOM FTP server (https://fantom.gsc.riken.jp/5/datafiles/reprocessed/hg38_latest/extra/CAGE_peaks/hg38_fair+new_CAGE_peaks_phase1and2.bed.gz) on June 15th 2020  [-@noguchi_fantom5_2017]. Transcriptional start sites (TSS) were extracted from gencode and DNTX annotations; TSS is defined as the start of the first exon of a transcript. Distance to CAGE peaks was calculated using the closest tool in the bedtools suite. Significant difference in mean distance to CAGE peak between DNTX and gencode annotation was tested with a Mann Whitney U test. 

|              Polyadenylation signal annotations were downloaded from the polyA site atlas  (https://polyasite.unibas.ch/download/atlas/2.0/GRCh38.96/atlas.clusters.2.0.GRCh38.96.bed.gz) on June 15th 2020 [-@herrmann_polyasite_2020]. Transcriptional end sites(TES) were extracted from gencode and DNTX annotations; TES is defined as the end of the terminal exon of a transcript. Distance to polyA signal was calculated using the closest tool in the bedtools suite [-@quinlan_bedtools_2010]. Significant difference in mean distance to polyA signal was tested with a Mann Whitney U test. 

## Identification of novel protein coding transcripts

|              Protein-coding transcripts in the unified transcriptome were identified using the TransDecoder suite [-@haas_novo_2013]. Transcript sequences in fasta format were extracted from the final pan-body transcriptome using the TransDecoder util script "gtf_genome_to_cdna_fasta.pl". Potential open reading frames(ORFs) were generated from transcript sequences using the LongestORF module within TransDecoder, and the single best ORF for each transcript was extracted with the Predict module within Transdecoder. The resulting ORFs were mapped to genomic locations with the TransDecoder util script "gtf_to_alignment_gff3.pl". For each ORF start and stop codons were extracted with the script "agat_sp_add_start_stop.pl" scripts from the AGAT toolkit (https://github.com/NBISweden/AGAT/). Transcripts with no detectable ORF or missing a start or stop codon were labelled as noncoding.

## Analysis of novel isoforms in eye tissues

|              An Upset plot was generated using the ComplexUpset package (https://github.com/krassowski/complex-upset) [-@lex_upset_2014]. Fraction Isoform Usage (FIU) was calculated for each transcript *t* associated with a parent gene *g* using the following formula: $FIU_t = \frac{TPM_t}{TPM_g}$ . Raincloud plots of FIU were generated using the "R_Rainclouds" R package  [-@allen_raincloud_2019].

## Analysis of fetal retina RNA-seq data.

|              RNA-seq samples from Mellough et al. were obtained from EiaD, and were not included in the main dataset used for building transcriptomes. Outliers within the dataset were identified by first performing principal component analysis of transcript level expression data, calculating the center of all data using the first two principal components, and subsequently removing five samples furthest away from the center of all data. The remaining samples were normalized using calcNormFactors from the R package edgeR and converted to weights using the voom function from the R package limma  [-@robinson_edger_2010], [-@ritchie_limma_2015]. Differential expression was modeled using the lmFit function using developmental time point as the model design and tested for significant change in expression using the Ebayes function from limma. Gene Set enrichment was tested using the R package clusterprofileR [-@yu_clusterprofiler_2012]. Heatmaps were generated using the ComplexHeatmap package [-@gu_complex_2016].

## Prediction of variant impact using *de novo* transcriptomes.

|              Noncoding variants previously associated with retinal disease from the Blueprint Genetics Retinal dystrophy panel were obtained from the Blueprint Genetics website (https://blueprintgenetics.com/tests/panels/ophthalmology/retinal-dystrophy-panel/). The variants were converted from HGVS to VCF format using a custom python script "HGVS_to_VCF.py". This VCF was then remapped to the hg38 human genome build using the tool crossmap [-@zhao_crossmap_2014]. The VCF of variants was used as the input variants for the Variant Effect Predictor(VEP) tool from Ensembl, with each subtissue specific transcriptome as the input annotation [-@mclaren_ensembl_2016]. VEP was additionally run using the gencode v28 comprehensive annotation as the input annotation to identify variants whose predicted impact increased in severity.

## Figures, Tables, and  Computing Resources

|              All statistical analyses, figures and tables in this paper were generated using the R programming language. [-@r_core_team_r_2019] A full list of packages and versions can be found in the supplementary file session_info.txt. All computation was performed on the National Institutes of Health high performance computer system Biowulf (hpc.nih.gov).

# Supplemental Figures

```{r sup_fig1 , include=T}
ggplot(flnc_lengths)+
 geom_violin(aes(x=lib, y=V2))+
 ylab('Transcript Length')+
 xlab('PacBio Library Type')+
 #ggtitle('Length Distribution of Full Length Non-Chimeric(FLNC) long-reads')+
 cowplot::theme_cowplot()
```

::: {custom-style="CustomCaption"}
Supplemental Figure 1. Distribution of PacBio long-read lengths for two library sizes.
:::

```{r sup_fig2, fig.height=6, fig.width=8, include = T, echo = F}

ym <- 10000
cl <- c('gencode' = 'blue', 'dntx' ='purple')
polya <- ggplot(all_polya_closest %>% rename(build=label)) +
 geom_boxplot(aes(x=build, y = abs_dist, color = build))+
 ylim(c(0,ym) )+
 ylab('TES distance to polyA signal')+
 xlab('')+
 scale_color_manual(values = cl)+
 coord_flip()+
 cowplot::theme_cowplot()

cage <- ggplot(all_CAGE_phase12) +
 geom_boxplot(aes(x=build, y = abs_dist, color = build))+
 ylim(c(0, ym) )+
 ylab('TSS distance to CAGE peak')+
 xlab('')+
 scale_color_manual(values = cl)+
 coord_flip()+
 cowplot::theme_cowplot()
phylo <- ggplot(data = all_phylop)+
 geom_boxplot(aes(x=build, y=mean_phylop_score, color = build))+
 ylab('Mean PhyloP score per exon')+
 xlab('')+
 ylim(0,1)+
 scale_color_manual(values = cl)+
 coord_flip()+
 cowplot::theme_cowplot()



all_val_plots <-  phylo / cage / polya 

all_val_plots +plot_layout(guides = 'collect') 
```

::: {custom-style="CustomCaption"}
Supplemental Figure 2. Comparison of DNTX annotation to GENCODE annotation. A) Average per exon Phylop score for GENCODE and DNTX transcripts. B) Average distance of DNTX transcriptional start sites (TSS) and GENCODE TSS to CAGE-seq peaks from the FANTOM consortium. C) Average distance of DNTX transcriptional end sites (TES) and GENCODE TES to polyadenylation signals in the PolyA site atlas.
:::

```{r map_rate_diff_prep, fig.height=5, fig.width=6}
map_rate_diffs <- all_sample_mapping_rate_difference %>%
 filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
 mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location=case_when(body_location == 'Body' ~ 'Body(mean)',
                 body_location == 'Brain'~ 'Brain(mean)' ,
                 TRUE ~ body_location),
     body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location_pretty=gsub('avg','mean', body_location_pretty),
     body_location_pretty=gsub('Tissue','', body_location_pretty),
     body_location_pretty=gsub('\\(',' (', body_location_pretty),
     body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
     body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty),
     pt = 'mapping rate reduction')

size_reduction_df <- inner_join(gencode_tx_size, tx_counts) %>%
 mutate(txome_size_decrease = ( all_exp-filtered) ) %>%
 inner_join(subtissue_to_bodyloc) %>%
 filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
 mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location=case_when(body_location == 'Body' ~ 'Body(mean)',
                 body_location == 'Brain'~ 'Brain(mean)' ,
                 TRUE ~ body_location)) %>% 
 mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
     body_location_pretty=gsub('avg','mean', body_location_pretty),
     body_location_pretty=gsub('Tissue','', body_location_pretty),
     body_location_pretty=gsub('\\(',' (', body_location_pretty),
     body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
     body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
 mutate(pt = '% Base txome removed', 
     txome_size_decrease = txome_size_decrease *-1)


```
```{r map_rate_diff, fig.height=5, fig.width=6, include = T}
cl<- novel_transcripts_per_tissue$color
names(cl) <- novel_transcripts_per_tissue$body_location
map_rate_diffs <- map_rate_diffs %>% mutate(diff = DNTX_percent_mapped - gencode_percent_mapped)
map_rate_d_plot <- ggplot(map_rate_diffs) +
 geom_boxplot(aes(x = pt, y = mapping_rate_diff)) +
 xlab('') + 
 ylab('Per-sample change in Salmon mapping rate\n(DNTX-gencode)') +
 cowplot::theme_cowplot()+
 coord_flip() +
 theme(axis.text.y=element_blank(), panel.background = element_blank())
 
 
 
frac_rm_plot <- ggplot(data = size_reduction_df) + 
  geom_boxplot(aes(x = pt, y=txome_size_decrease) ) +
  xlab('') +
  cowplot::theme_cowplot() +
  theme(axis.text.y=element_blank(), panel.background = element_blank()) +
  ylab('Per-subtissue reduction in Transcriptome Size\n(DNTX-gencode)') +
  theme(legend.position = "none")+
  coord_flip() 

frac_rm_plot / map_rate_d_plot + plot_annotation(tag_levels = 'A')

```

::: {custom-style="CustomCaption"}
Supplemental Figure 3. Comparison of Salmon mapping rate change vs transcriptome size decrease. 
:::


```{r vep_prep }
# pretty_subtissue <- function(st){
#  bodyloc_pretty <- st
#  bodyloc_pretty = str_replace_all(bodyloc_pretty, '_|\\.', ' (')
#  bodyloc_pretty= str_remove_all(bodyloc_pretty, ' \\(Tissue')
#  bodyloc_pretty = str_replace_all(bodyloc_pretty, '-', ' ')
#  bodyloc_pretty =replace(bodyloc_pretty, st%in% eye_tissues,
#                  paste0(bodyloc_pretty[ st%in% eye_tissues], ')') )
#  return(bodyloc_pretty)
# }
# 
# 
# draw_all_transcripts_interactive_v3 <- function(gene, gtf, tissues,keep_tx, var_df){
#   coldf <- inner_join(subtissue_to_bodyloc, tissue_color_mapping_df) %>% mutate(Subtissue = pretty_subtissue(subtissue))
#   cl <- coldf$color
#   names(cl) <- coldf$Subtissue
#   gtf_gene <- filter(gtf, gene_name == gene, transcript_id %in% keep_tx)
#   maxchar <- max(nchar(gtf_gene$transcript_id), nchar(gtf_gene$gene_name), 
#          nchar(gtf_gene$oId %>% na.omit %>% . [grepl('ENST', .)] ))
#   
#   plot_data <- gtf_gene %>% filter( type == 'exon') %>% 
#     mutate(`exon type`=ifelse(is.na(novel_exon_id), 'ref', 'novel')) %>% 
#    left_join(var_df) %>% 
#    left_join(ctab_plotting) %>%
#    left_join(coldf) %>% 
#    mutate(transcript_id = factor(transcript_id, levels = keep_tx))
#   set.seed(234)
#   intron_gaps <- tibble(Xmin=plot_data$Xmax[1:(nrow(plot_data) -1)], Xmax=plot_data$Xmin[2:(nrow(plot_data))]) %>% 
#     distinct %>% 
#     mutate(midpoint=rowMeans(.)) %>% 
#     filter(midpoint >0) %>% sample_n(2)
# 
#   point_type = ifelse(plot_data$strand[1] == '+',62, 60)
#   plot <- ggplot(data = plot_data) +
#     geom_hline(yintercept = 0, colour='black', size=.5)+
#     geom_point(data=intron_gaps, aes(x=midpoint),y=0,size=4, fill='black', shape=point_type)+
#     geom_rect(aes(xmin=Xmin, xmax=Xmax, ymin=Ymin, ymax=Ymax,fill=Subtissue))+
#     #geom_rect( aes(xmin=Xmin, xmax=Xmax, ymin=Ymin, ymax=Ymax,fill=`exon type`))+
#     scale_fill_manual(values = cl) +
#     geom_point(aes(x=var_x, y=var_y),fill = 'black', shape='*', size = 5)+
#     facet_wrap(~transcript_id, ncol=1, strip.position = 'right')+
#     #ggtitle(gene) +
#     theme_void() 
#     # theme(strip.text = element_text(angle = 180))
#   return(plot)
# }
# vep_gms <- draw_all_transcripts_interactive_v3(gtf = pg_ctx, gene = "CACNA2D4", keep_tx = all_tx_cleanids, var_df = var_plot) 
# 
# cl <- c('Low' = 'lightblue', 'Moderate' = 'lightgreen', 'High' = 'red')
# shp <- c('dntx'=16, 'gencode'=15)
# k <- vep_plot_df %>% 
#  filter(var_id %in% example_var_ids) %>% 
#  filter(!as.character(subtissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
#  left_join(subtissue_to_bodyloc) %>%
#  mutate(body_location = replace_na(body_location, 'gencode'),
#     body_location = factor(body_location, levels = c('gencode', eye_tissues, 'Body', 'Brain'))) %>%
#  group_by(body_location, var_id, build) %>% 
#  summarise(`avg log2(TPM + 1)` = mean(`log2(TPM + 1)`),
#       `Variant Impact` = `Variant Impact`[which.max(max_impact)],
#       highest_max_impact =max(max_impact)) %>% 
#  arrange(body_location) %>% 
#  mutate(bodyloc_pretty = as.character(body_location), 
#     bodyloc_pretty = pretty_subtissue(bodyloc_pretty),
#     bodyloc_pretty = factor(bodyloc_pretty, levels = bodyloc_pretty))
#       
# 
# vep_pred <- ggplot(k, aes(x =var_id, y=bodyloc_pretty))+
#  geom_point(aes(size = `avg log2(TPM + 1)`, color = `Variant Impact`, shape=build))+
#  scale_color_manual(values = c('Low'='mediumpurple', 'Moderate'='green', 'High'='tomato'))+
#  scale_shape_manual(values = shp)+
#  geom_hline(yintercept = 1.5)+
#  ylab('')+
#  xlab('ClinVar Variation ID')+
#  ggtitle('Tissue Specific Variant Prioritization') +
#  cowplot::theme_cowplot() +
#  coord_flip()+
#  theme(axis.text.x.bottom = element_text(angle = 45, hjust = 1), panel.background = element_blank())

```
```{r vep, fig.width=12,fig.height=7}
# vep_pred / vep_gms +plot_annotation(tag_levels = 'A')
# ::: {custom-style="CustomCaption"}
# Figure 5. Tissue specific variant prioritization of two variants of unknown significance (VUS) previously associated with ocular disease. A) Variant Impact dot plot. Size corresponds to average TPM of the transcript with the highest TPM and color to impact. A full mapping of variant descriptors to impact can be found in supplemental data. b) Transcript Models of CACNA2D4 thick lines represent CDS; Asterisk represents location of variant 631715. Arrows indicate direction of transcription
# :::
# 
# We compared our tissue specific variant effect predictions to variant effect predictions generated using the gencode transcript annotation. We found that globally out of `r toString(NUM_TOTAL_VUS)` total variants examined, the priority of `r toString(NUM_IMPACT_DECREASE)` variants decreased in at least one tissue, and `r toString(NUM_IMPACT_INCREASE)` variants increased in at least one tissue. We see a larger amount of variants decrease in predicted effects as the transcript which previously lead to an effect is not expressed in a particular tissue. To demonstrate the utility of tissue specific variant prioitizaion, we focused specifically on variants previously associated with ocular disease. We examined ClinVar variant 631672, which was previously associated with retinal cone dystrophy 4, and has a high predicted impact only in ocular tissues, and not in other tissues. (Fig 5A) Variant 631672 is in the gene CACNA2D4 and in all transcripts expressed in ocular tissues this variant leads to a premature stop codon. While CACNA2D4 has other isoforms expressed in other tissues, these isoforms are not affected by variant 631672 and thus there is a higher probability that this variant may be playing a role in disease.(Fig 5B) Conversely, in clinvar variant 631715 which is associated with posterior column ataxia-retinitis pigmentosa syndrome and is within gene FLVCR2, the variant is low impact on transcripts expressed in eye tissues, but higher impact in other body tissues. This may suggest that this variant is less likely to be associated with the disease.
```
```{r, primary_isoforms_prep, eval=F}
## *De novo* transcriptomes identify of the tissue specific primary isoform expressed in a tissue
#    It is useful researchers to know which isoform of a gene is the most relevant to a particular biological system. The APPRIS project identified principal isoforms for humans, based on predicted function and evolutionary conservation, using multiple computational tools. Principal isoforms are considered the most phenotypically relevant isoforms of transcripts. We instead propose a tissue specific primary isoforms, defined as the highest expressed isoforms of a given gene in a given tissue. As primary isoforms are based on biological evidence via RNA-seq, we reason that it is more refined than APPRIS principal isoforms.
# primary_isoforms_per_tissue <- primary_isoforms_per_tissue %>%
#  mutate(primary_isoform_origin = gsub('dntx','DNTx', primary_isoform_origin),
#     primary_isoform_origin = gsub('gencode', 'GENCODE', primary_isoform_origin),
#     primary_isoform_origin = gsub('appris', 'APPRIS', primary_isoform_origin)) %>% 
#  mutate(primary_isoform_origin = factor(primary_isoform_origin, levels = c('DNTx', 'GENCODE', 'APPRIS'))) %>% 
#  inner_join(subtissue_to_bodyloc) %>%
#  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
#  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
#     body_location_pretty=case_when(body_location == 'Body' ~ 'Body(mean)',
#                     body_location == 'Brain'~ 'Brain(mean)' ,
#                     TRUE ~ body_location)) %>% 
#  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
#     body_location_pretty=gsub('avg','mean', body_location_pretty),
#     body_location_pretty=gsub('Tissue','', body_location_pretty),
#     body_location_pretty=gsub('\\(',' (', body_location_pretty),
#     body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
#     body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
#  group_by(body_location_pretty, primary_isoform_origin) %>%
#  summarise(count = mean(count)) %>% 
#  arrange()
# ggplot(data = primary_isoforms_per_tissue, aes(x = body_location_pretty, y=count, fill = primary_isoform_origin))+
#  geom_bar(position = 'fill', stat = 'identity')+
#  geom_text(aes(label = as.integer(count) ), position = position_fill(vjust=.5)) +
#  cowplot::theme_cowplot() +
#  theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank()) +
#  scale_fill_brewer(palette = 'Set1', name = 'Origin') + ylab('Ratio') + xlab('')
# 
# ```
# ```{r primary_isoforms, fig.width= 6, include=T}
# ggplot(data = primary_isoforms_per_tissue, aes(x = body_location_pretty, y=count, fill = primary_isoform_origin))+
#  geom_bar(position = 'fill', stat = 'identity')+
#  geom_text(aes(label = as.integer(count) ), position = position_fill(vjust=.5)) +
#  cowplot::theme_cowplot() +
#  theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank()) +
#  scale_fill_brewer(palette = 'Set1', name = 'Origin') + ylab('Ratio') + xlab('')
# 
# ::: {custom-style="CustomCaption"}
# Figure 4. APPPRIS principal isoforms within the *de novo* transcriptome
# :::
# 
#    {== We found that `r toString(NUM_APPRIS_RECOVERY_RATE)` of the APPRIS principal isoforms are also identified as the primary isoform in at least one tissue type. Within each tissue type, on average `r toString(NUM_AVG_APPRIS)` % of are an appris primary isoform, `r toString(NUM_AVG_NONAPPRIS_GENCODE)` of transcripts were a previously annotated gencode transcript, `r toString(NUM_AVG_NOVEL)`
# 
# \% of principal isoforms are novel isoforms. ==}{>> need to fix these numbers <<}

# {== We initially reconstructed `r toString(round(NUM_INITIAL_ANOO_TX/NUM_TOTAL_GENCODE_TX*100), digits = 3)` % of the Gencode transcriptome, but after filtering we were left with `r toString(round(NUM_REF_TX_BODY/NUM_TOTAL_GENCODE_TX*100), digits = 3)` %. This was expected as similar results are seen in the CHESS database with which we share some of our underlying data with. This suggests that some of the Gencode transcripts may not be relevant to every human organ system, and that the one size fits all strategy of having a single comprehensive annotation for all human tissues may not be the most optimal one.(still need to DO SOME LITERATURE SEARCH TO SEE IF THIS IS MENTIONED ANYWHERE. GTEx papers?). This is further supported by the minimal loss loss in gene expression data despite the large decrease in transcriptome size for each tissue. This reduction in total transcriptome sizes is also useful for other researchers as it creates a smaller search space to explore, ie it is easier to compare 30,000 transcripts than 300,000 ==}{>> Maybe cut this paragraph? <<}

# ## Novel Loci in the ocular transcriptome
# 
# We determined the longest ORF’s for each novel loci, and for each transcript with a valid ORF, predicted the function the encoded protein using hmmer searching within the pfam databse, and blastp searching within the swissprot database. We chose to hmmer results that had a sequence match e-value < 1e-4, and blastp results with e-value < 1e-15 (Sup Table). We found XXX distinct potentially novel coding loci from blast and hmmer combined. We highlight one detected in fetal RPE
# library(patchwork)
# 
# cov <- ggplot(cov_df, aes(x=coord,y=cov, group = sample))+
#  geom_line( alpha = .3)+
#  geom_area(alpha = .3)+
#  #ggtitle(paste(txid, s_st))+
#  theme(axis.title.x = element_blank(), panel.background = element_blank(), 
#     axis.text.x = element_blank(), axis.ticks.x = element_blank())
# pt <- ifelse(gene_model_df$strand[1] == '+',62, 60)
# model <- ggplot(gene_model_df)+
#  geom_rect(aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax))+
#  geom_rect(xmin=min(gene_model_df$xmin), xmax=max(gene_model_df$xmax), ymin=-.1, ymax=.1)+
#  geom_point(data=point_df, aes(x = x, y=y), shape= pt, size = 5)+
#  xlab('genome') +
#  ylab(element_blank())+
#  theme(panel.background = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank())
# layout <- '
# AA
# AA
# AA
# AA
# AA
# AA
# BB
# CC
# CC
# '
# mss <- as_ggplot(rasterGrob(readPNG('clean_data/plots/msa_ex.png'), interpolate = T))
# cov/model/mss + plot_layout(design = layout) +plot_annotation(tag_levels = 'A')
# 
# ::: {custom-style="CustomCaption"}
# Figure 6. DNTX000dddd, potentially protein coding novel isoform in Fetal RPE. A) RNA-seq coverage of exons across 49 samples. B transcript model of DNTX000ddd. thick regions represtn coding region. C. Multiple sequence alignment of DNTX000ddd and closely related proteins.
# :::
# 
# Placeholder figure, still trying to find the right transcript to look at here.


#### Discussion overFlow

#We see many more distinct transcripts in the long-read data than in the short were not corroborated in our short-read based transcriptomes. This low concordance could be to technical error arising from the ability of StringTie to accurately build transcripts. However, This low concordance could simply be transcriptional noise, as others have suggested. An approach that uses many long-read based transcriptomes across many samples would help to disentangle what is noise and what is signal; at this time this would be prohibitively expensive as long-read RNA-seq is approximately five to ten times the cost of short-read RNA-seq.

```



# References
