---
title: '*De novo* transcriptomes built from hundreds of eye tissues reveal hundreds of novel gene isoforms'
author: 'Vinay S Swamy, David McGaughey'
bibliography: 'ocular_txome_citations.bib'
csl: 'nucleic-acids-research.csl'
output:
  word_document: 
            reference_docx: reference_doc_v1.docx  
            keep_md: yes
  html_document:
    df_print: paged
---

```{r setup, echo=FALSE}
#knitr::opts_knit$set(root.dir = '~/NIH/occular_transcriptomes_paper/')
knitr::opts_knit$set(root.dir = '/Volumes/data/ocular_transcriptomes_paper/')
#knitr::opts_knit$set(root.dir = '/data/swamyvs/occular_transcriptomes_paper')
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```


```{r prep}
library(tidyverse)
library(DT)
library(RColorBrewer)
library(UpSetR)
library(ComplexHeatmap)
library(ggpubr)
library(knitr)
library(grid)
library(png)
sample_table <- read_tsv('sampleTableFull.tsv') %>% filter(subtissue != 'synth')
load('clean_data/rdata/tissue_to_colors.Rdata')
load('clean_data/rdata/core_tight.Rdata')
tissue_color_mapping_df <- bind_rows(tissue_color_mapping_df,  tibble(body_location=c('Brain(avg)', 'Body(avg)'), color=c('orange','yellow')))
load('clean_data/rdata/transcriptome_pipeline_stats.Rdata')
subtissue_to_bodyloc <- sample_table %>% select(subtissue, body_location) %>% distinct
core_tight <- core_tight %>% select(sample=sample_accession, study=study_accession) %>% distinct
load('clean_data/rdata/paper_numbers.Rdata')
```


::: {custom-style="anchor"}
1-open
:::

# Introduction

|        The transcriptome is defined as the set of distinct mRNA transcripts expressed in a population of identical cells. Although there are an estimated 20,000 genes, during transcription several RNA processing steps modify immature RNA and drive the formation of multiple, distinct gene isoforms.  This processing includes alternative promoter usage, alternative splicing, and alternative polyadenylation. The full biological impact of gene isoforms has not been fully elucidated, but several studies have shown that gene isoforms can have distinct biological functions.
       The detection of novel gene isoforms is a complicated task, but RNA-seq has been used as a powerful tool for investigating gene isoforms. Early methods using RNA-seq to detect gene isoforms generally focused solely on investigating the single RNA processing mechanism and determining what gene isoforms it was driving. For example, the computational tool rMATS detects novel gene isoforms by analyzing splicing patterns using RNA-seq[-@shen_rmats_2014]. More recently, researchers have adapted methods from genome assembly to use RNA-seq to reconstruct the whole transcriptome of a biological samples, dubbed *de novo* transcriptome construction.[-@haas_novo_2013]
|        *De novo* transcriptome construction uses short RNA-seq reads to reconstruct full-length mRNA transcripts. This is a particularly useful method because it is independent of any RNA processing mechanism. A major shortcoming of this method is that a large number of samples is required to combat the noisy nature of RNA-seq data, but because of the increasingly inexpensive sequencing, data sets of the necessary size are now available. The most comprehensive *de novo* transcriptome project to date has been CHESS, which used the GTEx dataset to construct *de novo* transcriptomes in over 9000 RNA-seq samples from 49 distinct location of the body to create a comprehensive annotation of mRNA transcripts across the human body. [-@pertea_stringtie_2015],[-@gtex_consortium_genetic_2017],[-@pertea_chess_2018] However, as the GTEx dataset lacks any ocular tissues, the CHESS database is an incomplete annotation of the human transcriptome
       In this study we *de novo* constructed the transcriptome of the eye using a previously published data set containing over 800 ocular tissue samples compiled from mining publicly available sequencing data. [@swamy_eye_2019] ) We focused on three major ocular subtissues: The cornea, retina, and the retinal pigmented epithelium (RPE) using RNA-seq data from both adult and fetal tissues and used rigorous analysis pipeline to create a high-quality pan-eye transcriptome, revealing hundreds of novel gene isoforms as well as several novel genes. In addition to our ocular samples, we used a subset of the GTEx dataset to construct transcriptomes for 49 other locations across the body to facilitate comparisons in transcriptomes across the body. Here, we provide our *de novo* transcriptomes as a resource to other researchers through an R package and webapp.

::: {custom-style="anchor"}
1-close
:::

# Methods 

## Construction of initial tissue specific transcriptomes.

::: {custom-style="anchor"}
2-open
:::

|        We identified studies with healthy, unperturbed RNA-seq samples from `r toString(NUM_BODY_TISSUES + 6)` distinct subtissue regions of the body, downloaded and performed quality control the pertinent sequencing data from the sequence read archive (SRA) using methods from [@swamy_eye_2019] We aligned each sample to the Gencode V28 hg38 assembly using the genomic aligner STAR and sorted the resulting BAM files using samtools sort.[-@frankish_gencode_2019],[-@dobin_star_2013],[-@li_sequence_2009] For each sorted BAM file, we constructed a per sample base transcriptome using stringtie with the Gencode v28 comprehensive annotation as a guiding annotation and default parameters.[-@frankish_gencode_2019],[-@pertea_stringtie_2015] Next, for each base transcriptome we used stringtie merge to remove transcripts with Transcript-per-Million(TPM ) expression level  < 1 and minimum isoform fraction < .01. The resulting filtered base transcriptomes were merged the subtissue level using gffcompare (default parameters) with the Gencode V28 comprehensive annotation as the reference, to create 52 distinct tissue specific transcriptomes.

## Refinement of tissue-specific transcriptomes by transcript construction frequency

|        For each tissue-specific transcriptome, we used the transcript tracking file generated by gffcompare in the previous step to identify the number of samples from which a given transcript was constructed. We kept transcripts that were detected in samples from at least 3 different studies. For subtissues with fewer than 3 different studies, we kept transcripts constructed in at least 3 samples across all available studies.  At this stage we also removed novel loci that overlapped any region within 5 kilobases of a known gene by using bedtools intersect.[-@quinlan_bedtools_2010] Additionally, we obtained a bed file of known genomic repeat regions from the UCSC table browser and removed any transcripts that overlapped repeat regions.[-@haeussler_ucsc_2019]

## Refinement of tissue-specific transcriptomes by transcript quantification
|        For each resulting filtered transcriptome, we extracted transcript sequences using the tool gffread, and used these sequences to build a tissue-specific quantification index using the index mode of the alignment free quantification tool Salmon using the following parameters: --type quasi --perfectHash --k 31.[-@patro_salmon_2017] For each sample, we quantified transcript expression using the quant mode of salmon, using a samples respective tissue specific quantification index with the following runtime parameters: --gcBias --seqBias --numBootstraps 100 --validateMappings.  For, for each tissue-specific transcriptome, we used salmon’s bootstrap quantification data to calculate the quantification variance for each transcript. We used this variance calculation to generate a distribution of transcript quantification variance for previously annotated transcripts, and then removed novel transcripts with a quantification variance greater than the 95th percentile of the reference transcript quantification variance. We additionally removed transcripts that had zero counts across all samples. This procedure  produced a final set of tissue-specific transcriptomes. These 52 tissue-specific transcriptomes were merged into a single unified transcriptome using gffcompare using --strict-match mode, and DNTX as the transcript prefix.

## Identification of protein coding novel transcripts. 

|        We identified protein coding transcripts in our unified transcriptome using the TransDecoder suite.[-@haas_novo_2013] We extracted transcript sequences using the util script gtf_genome_to_cdna_fasta.pl and used TransDecoder to find a single best open reading frame from each transcript. Transcripts with no detectable open reading frame were considered noncoding. We used TransDecoder utility script cdna_alignment_orf_to_genome_orf.pl to create a gff3 annotation for all protein coding transcripts.

## Calculation of PSI for novel exons.

|        We define percent-spliced  -in (PSI) as the following:

$PSI = \frac{\Sigma\ inclusion\ reads}{\Sigma\ inclusion\ reads\ +\ \Sigma\ inclusion\ reads}$

where inclusion reads are all reads spanning an exon’s splice junctions, and exclusion reads are all reads spanning the any exon-exon junction that excludes a given exon. For each sample, we calculated the inclusion and exclusion reads using the tool rMATS, using each tissue-specific transcriptome as the annotation for its respective tissues, and reference index built against hg38.[-@shen_rmats_2014] We determined total inclusion and exclusion reads for a given exon by summing all inclusion and exclusion reads for all distinct alternative splice events associated with a given exon and used these to calculate PSI.

## Computing Resources
|        All computation was performed on the National Institutes of Health cluster compute system Biowulf.

## Code availability.
|        To improve reproducibility, we wrote all code used to generate both the data and figures for this paper as Snakemake pipelines.[-@koster_snakemakescalable_2012] All code used for this pipeline is publicly available  in the github repository https://github.com/vinay-swamy/ocular_transcriptomes_pipeline, and all code to produce these figures is available in the github repository https://github.com/vinay-swamy/ocular_transcriptomes_paper

## Figures and Tables
|        All statistical analyses, figures and tables were generated using the R programming language.[-@r_core_team_r_2019] A full list of packages and versions can be found in supplementary file session_info.txt

::: {custom-style="anchor"}
2-close
:::

# Results 


## Long Read Pacbio RNA sequencing guides *de novo* transcriptome construction
In order to determine the accuracy of short read transcriptome construction, we first generated PacBio long read RNA-seq data from a stem cell derived RPE cell line.  



## A rigorous analysis pipeline finds thousands of novel gene isoforms

```{r}
df <- sample_table %>% 
  left_join(core_tight) %>% 
  filter(!body_location %in% c('Body', 'Brain', 'ESC_Stem.Cell.Line', 'Lens_Stem.Cell.Line')) %>% 
  group_by(subtissue) %>%
  summarise(`number of samples` = n(), `number of studies` = length(unique(study))) %>% 
  arrange(desc(`number of studies`)) %>% 
  inner_join(tx_counts %>% select(subtissue, `transcriptome size`=final)) %>% 
  mutate(subtissue=gsub('_|\\.',' ', subtissue)) 
kable(df)
```

::: {custom-style="anchor"}
3-open
:::

::: {custom-style="CustomCaption"}
Table 1. Ocular sample dataset. Transcriptome size is defined as the number of unique transcripts expressed in a given tissue type
:::

|       We built transcriptomes from `r toString(NUM_EYE_SAMPLES)` previously published, publicly available ocular tissue RNA-seq samples using an efficient snakemake pipeline.(sup fig 1) We include both adult and fetal tissue from the Cornea, Retina, Retinal Pigmented Epithelium (RPE) mined from `r toString(NUM_STUDIES)` different studies (Table 1). Our fetal tissues consist of both human fetal tissues and human induced pluripotent stem cell (iPSC) derived tissue. To more accurately determine tissue specificity of novel ocular transcripts,  we supplemented our ocular data set with `r toString(NUM_BODY_SAMPLES)` samples across `r toString(NUM_BODY_TISSUES)` body locations from the GTEx project and constructed transcriptomes for each of these body locations. We refer to each distinct body location as a subtissue here after.
|               After initial construction of transcripts, we found over two million distinct transcripts across all our samples. Because we observed that many of these transcripts were detected in one or two samples (sup fig 2) we refined our set of transcriptomes using series of filtering steps (methods) to remove low confidence transcripts. Our final transcriptome across all tissues contains `r toString(NUM_TOTAL_TX_BODY)` distinct transcripts with `r toString(NUM_REF_TX_BODY)` previously annotated and `r toString(NUM_NOVEL_TX_BODY)` novel transcripts, and incudes `r toString(NUM_TOTAL_NOVEL_SEQ/1000000)` megabases of previously unannotated genomic sequence. We define novel as any region of genomic sequence that is not found in Gencode, Ensembl and Refseq annotation databases. [-@frankish_gencode_2019],[-@zerbino_ensembl_2018],[-@oleary_reference_2016] Novel transcripts are split into two categories: novel isoforms, which are novel variations of known genes, and novel loci, which are previously unreported, entirely novel regions of transcribed sequence.

::: {custom-style="anchor"}
3-close
:::

```{r transcriptome_pipeline_stats, fig.height=13, fig.width=16}
#  novel_loci bar graph, novel isoform bargraphs initial tx count, final tx count
load('clean_data/rdata/buildResultsSummary.Rdata')
novel_transcripts_per_tissue <- novel_transcripts_per_tissue %>% filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(body_location_pretty=gsub('_|\\.', ' ', body_location))
color_list<- novel_transcripts_per_tissue$color
names(color_list) <- novel_transcripts_per_tissue$body_location
isoforms <- ggplot(data = novel_transcripts_per_tissue) +
    geom_col(aes(x=body_location_pretty, y=novel_transcript_count, fill=body_location)) +
    scale_fill_manual(values = color_list)+
    ggtitle('Novel Isoforms Contructed Across the Body')+
    ylab('number of novel transcripts')+
    xlab('Body Location')+
    theme(axis.text.x=element_text(angle=45, hjust = 1))
novel_loci_per_tissue <- novel_loci_per_tissue %>% filter(transcript_type!='color',!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(counts=as.numeric(counts), body_location_pretty=gsub('_|\\.', ' ', body_location))

novel_exons <- as_ggplot(rasterGrob(readPNG('clean_data/plots/sankey_novel_exons.png'), interpolate = T))


loci <- ggplot(data = novel_loci_per_tissue) +
    geom_col(aes(x=body_location_pretty, y=counts, fill=body_location, alpha=transcript_type), position = 'dodge') +
    scale_fill_manual(values = color_list) +
    scale_alpha_discrete(range=c(.5,1)) +
    ggtitle('Novel Loci constructed in Occular Tissues', ) +
    ylab('number of novel loci detected') +
    xlab('Body Location\n\n\n') +
    theme(axis.text.x=element_text(angle=45, hjust = 1))

p0 <- ggarrange(isoforms, loci, ncol = 2, labels = c('A', 'B'), font.label = list(size=25))
p1 <- annotate_figure(novel_exons, fig.lab = 'C', fig.lab.pos = 'top.left', fig.lab.size = 25)
ggarrange(p0, p1, nrow=2)
```

::: {custom-style="anchor"}
4-open
:::

::: {custom-style="CustomCaption"}
Figure 1 (A, B) Number of novel gene isoforms and novel loci constructed across ocular and other body tissues. Brain and body represent an average of 13 and 34 distinct subtissues, respectively (C -) Classification of novel gene isoforms based on source of novelty.
:::

::: {custom-style="anchor"}
4-close
:::
  
```{r}
# load('/Volumes/data/ocular_transcriptomes_pipeline/data/rdata/novel_exon_classification.Rdata')
# gtf <- rtracklayer::readGFF('/Volumes/data/ocular_transcriptomes_pipeline/data/gtfs/all_tissues.combined_NovelAno.gtf')
# # 
# # gtf_exons <- gtf %>% filter(type == 'exon')
# omit_exons <- gtf %>% filter(type == 'exon', !grepl('ENST', oId)) %>% group_by(transcript_id) %>% summarise(num_novel_exon=sum(is.na(novel_exon_id))) %>% filter(num_novel_exon == 0) %>% pull(transcript_id)
# event_counts <-  novel_exons_TSES %>% pull(nv_type_rc) %>% table %>% {tibble(type=names(.), count=.)} %>% 
#   mutate(type = case_when( type == 'A3SS'| type == 'A5SS' ~ 'Alternative Splice Site', 
#                            type == 'novel_exon' ~ ' Fully Unannotated Exon',
#                            type == 'RI' ~ 'Retained Intron', 
#                            type == 'novel_TES' ~ 'Novel Transcriptional End Site', 
#                            type == 'novel_TSS' ~ 'Novel Transcriptional Start Site')) %>% group_by(type) %>% summarise(num_ex=sum(count)) %>% mutate(isoform_type = 'novel_exon') %>% select(isoform_type, type, num_ex)
# bind_rows(event_counts, event_counts %>% )
# sum(event_counts$num_ex)
# length(omit_exons)
# # ggplot(event_counts) + 
# #   geom_col(aes(x=type, y=count, fill=type)) + 
# #   theme_minimal()
# 
# kable(event_counts)
```

::: {custom-style="anchor"}
5-open
:::

|        Novel isoforms occur due to a novel arrangement of known exons or contain a novel exon, which in turn is either a modification of a known exon to generate a novel modified exon, or an addition of a previously unannotated exon, or fully novel exon. For both classes of novel exon, we further annotated them with the putative biological process driving their inclusion: alternative splicing, alternative promoter usage, or alternative polyadenylation. We found that the majority of novel exons with our dataset are novel first and last exons.

## *de novo* transcriptomes improve sample mapping rates 

::: {custom-style="anchor"}
5-close
:::

```{r, fig.width=8}

#load('/Volumes/data/ocular_transcriptomes_paper/clean_data/rdata/all_tx_mappingrates.Rdata')
#***^deleteme^*** after I run the snakemake pipeline again

# df <- sample_mapping_rates %>% 
#   left_join(subtissue_to_bodyloc, .) %>% 
#   group_by(body_location) %>% 
#   summarise(mean_med_diff=mean(med_diff)*100, mean_raw_tx_count=mean(raw), mean_final_tx_count=mean(final)) %>% 
#   arrange(mean_med_diff) %>% 
#   mutate(body_location=case_when(body_location == 'Body' ~ 'Body(avg)', 
#                                  body_location == 'Brain'~ 'Brain(avg)' , 
#                                  TRUE ~ body_location)) %>% filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(body_location_pretty=gsub('_|\\.', ' ', body_location))
#   
# ggplot(data=df) + 
#   geom_col(aes(x=factor(body_location,levels = body_location,  ordered = T),y=mean_med_diff, fill=body_location)) + 
#   scale_fill_manual(values = color_list) +
#   xlab('subtissue') + 
#   ylab('median change in Salmon mapping rate') +
#   ggtitle('Changes in mapping rate between tissue specific *de novo* \ntranscriptomes and gencode V28 quantification ')+
#   coord_flip() + 
#   theme_minimal()
  
# all_sample_mapping_rates <- all_sample_mapping_rates %>% 
#   filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
#   mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),                                                              body_location=case_when(body_location == 'Body' ~ 'Body(avg)', 
#                                  body_location == 'Brain'~ 'Brain(avg)' , 
#                                   TRUE ~ body_location))
# 
#   
# ggplot(data = all_sample_mapping_rates)  + 
#   geom_boxplot(aes(x=build, y=mapping_rate, color=body_location)) + 
#   scale_color_manual(values = color_list) +
#   facet_wrap(~body_location, nrow = 3, ) +
#   ylab('Salmon Mapping Rate') + 
#   xlab('Transcriptome build')

# k <- '
# 
# ::: {custom-style="CustomCaption"}
# Figure 3 Salmon pseudo-mapping rates using both tissue -specific *de novo* transcriptome and gencode transcriptomes as the underlying reference for salmon. 
# :::
# 
# |        As an initial benchmark of the accuracy and utility of our transcriptome builds, we quantified transcript expression of our samples using the alignment-free quantification tool Salmon. Each sample was quantified twice: once using the gencode V27 human transcript annotation, and once using its associated tissue specific transcriptome. We found that globally, Salmon mapping rates increased when using the *de novo* transcriptomes as the reference transcriptome for quantification, despite an average of 3 -fold reduction in annotation size. Using our transcriptomes, we quantified a total of 13420034 reads across all of our samples. Additionally, for most tissue types, using *de novo* transcriptomes led to a more precise estimation of gene expression.
# 
# '

```


## Novel Isoforms in Ocular tissues 

```{r, fig.width=10, fig.height=10}
load('clean_data/rdata/novel_isoforms.Rdata')
source('scripts/R_raincloud.R')
plot_list <- novel_eye_tx_by_tissue[!names(novel_eye_tx_by_tissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line') ]
names(plot_list) <- gsub('_|\\.', ' ', names(plot_list))
png('/tmp/upset.png', res=100, height = 1000, width = 2000)
upset(fromList(plot_list),nintersects = 20, nsets = length(novel_eye_tx_by_tissue), order.by = 'freq')
dev <- dev.off()

us <- as_ggplot(rasterGrob(readPNG('/tmp/upset.png'), interpolate = T))

### PIU
# p <- ggviolin(piu_df, x='stage', y='piu', #fill = 'tissue', 
#               trim=T, draw_quantiles = T,
#                #title ='Comparison of percent isoform usage(piu) of novel \ntranscripts in fetal and adult eye tissues', 
#               ylab = 'fraction of total gene expression')+
#     #stat_compare_means(label.y = 1.1) +
#     scale_fill_manual(values = c(Cornea='green', Retina='blue', RPE='red'))
# piu <- facet(p,facet.by = 'tissue') +
#     theme_minimal()
# piu
 #RAIN CLOUD BITCHES
piu_df <- piu_df %>% mutate(new_stage=ifelse(stage == 'adult', 'Adult', 'Fetal'), 
                            subtissue=paste0(tissue, '_', new_stage, '.Tissue'))
piu <- ggplot(data=piu_df, aes(x=stage, y=piu, fill=subtissue, color=subtissue)) + 
  geom_flat_violin(position = position_nudge(x = .2, y = 0),adjust = 2)+
  geom_point(position = position_jitter(width = .15), size = .25, alpha=.2) + 
  geom_boxplot(outlier.shape = NA, alpha = 1, width = .1, colour = "BLACK")+ 
  scale_fill_manual(values = color_list) +
  scale_color_manual(values = color_list) +
  ylab('piu')+xlab('tissue')+coord_flip()+ theme_minimal()+ facet_wrap(~tissue, nrow = 3)

### exon locations
locs <- ggplot(location_df) + 
    geom_bar(aes(x=age, fill=location, y=count), position = 'fill', stat = 'identity' ) +
    scale_fill_manual(values = c('yellow', 'purple', 'orange')) +
    facet_wrap(~ tissue) +  
    ylab('percentage of novel exons') #+ 
    #ggtitle('location of novel exons in occular tissues')

bottom <- ggarrange(piu, locs, nrow = , ncol = 2, labels = c('B', 'C'))

ggarrange(us, bottom, nrow = 2, ncol = 1, labels = 'A' )

```

::: {custom-style="anchor"}
6-open
:::

::: {custom-style="CustomCaption"}
Figure 2. Analysis of novel isoform characteristics (A.). Set intersection of novel isoforms in ocular transcriptomes (B. Raincloud plots of novel isoform PIU (percent isoform usage). Boxplots are overlaid over PIU data with estimated distribution of data set above each boxplot (C.). Location of novel exons within the body of associated parent transcript. CDS refers to coding region of transcript, UTR to untranslated region and NC to noncoding transcript
:::

Next, we analyzed the novel isoforms within our ocular transcriptomes. We compared the overlap in constructed novel isoforms across ocular tissues and found that 95 % of novel isoforms are specific to a singular ocular subtissue (fig 3a). For each novel isoform we then calculated percent isoform usage (PIU), or the fraction of total gene expression a transcript contributed to its parent gene. We found that on average for retina and RPE, novel isoforms contribute to 30 % of their parent gene’s expression, whereas in the cornea we observed that novel isoforms contributed on average to 75% their parent gene’s expression.
In order to determine whether the novel exon in the novel isoform caused a change in the protein coding region of the transcript, we next found the longest open reading frame for each novel isoform with a novel exon. Novel isoforms with no detectable open reading frame had associated novel exons marked as noncoding. We found that novel exons lie largely in the untranslated regions of the transcript body. Less than 10% of detected novel isoforms potentially cause a change in its translated protein.

::: {custom-style="anchor"}
6-close
:::

## A companion visualization tool enables easy use of *de novo* transcriptomes 

```{r, fig.height=6}
piu <- as_ggplot(rasterGrob(readPNG('clean_data/plots/piu_shiny_app.png'), interpolate = T))
frac_det <- as_ggplot(rasterGrob(readPNG('clean_data/plots/frac_sample_det.png'), interpolate = T))
gene_body <- as_ggplot(rasterGrob(readPNG('clean_data/plots/gene_bodies_shiny_app.png'), interpolate = T))
p0 <- ggarrange(piu, frac_det, nrow = 1, labels = c('', 'C'))
ggarrange(p0, gene_body,  ncol =1, labels = c('A', 'B'))

```

::: {custom-style="anchor"}
7-open
:::

::: {custom-style="CustomCaption"}
Figure 3. Screenshots from dynamic *de novo* transcriptome visualization tool. A).  PIU bar plot for selected gene and tissue. B). Exon level diagram of transcript body Thicklines represent coding region of transcript. novel exons colored in red. Tooltip contains genomic location and phylop score C) Bargraph of fraction of samples within dataset each transcript was consructed in by tissue.
:::

|        To make our results easily accessible, we designed a webapp for visualizing and accessing our *de novo* transcriptomes. Users start by selecting gene or search for a gene by genomic location, and selecting up to 5 tissues to visualize transcript expression in. For each tissue we show the PIU for each transcript associated with a gene. We show the exon-intron structure of each transcript and mousing over exons show genomic location overlapping SNPs, and phylogentic conservation score. We additionally show a barplot of the fraction of samples in each tissue each transcript was constructed in. Users can also download the *de novo* transcriptomes for selected tissues in GTF format. 

# Discussion

|       Motivated by the lack of a comprehensive pan eye transcriptome and so we created the first comprehensive set of ocular transcriptomes. We used a large dataset compiled from published RNA-seq data to build these transcriptomes, an approach which has several key advantages. First, our large sample size allows us to combat the noisy nature of RNA-seq data. More importantly, because we only keep transcripts detected in samples from multiple studies with multiple types of sample preparation, we can be more confident that our transcriptomes accurately reflect the biology of its originating subtissue and are not an artifact due to preparation of the samples.
      We observed that in the set of novel exons within our transcriptome, the vast majority are novel first and last exons. It is difficult to directly tell what the biological relevance of these might be. There are multiple studies examining the roles of alternative first and last exons, with first exons arising because of alternative promoter usage, and last exons by alternative polyadenylation.[-@demircioglu_pan-cancer_2019],[-@mitra_alternative_2018] These studies have highlighted the distinct biological role these phenomena play. However, other studies have shown that many of these differences in start sites are largely nonfunctional and lack biological significance.[-@xu_evidence_2019] Within our data set we can only report the existence of these novel exons; a more intensive analysis of novel first and last exons is required to better identify their biologically relevance.
      We show that across all tissue types, the number of constructed transcripts is dramatically less than the number of transcripts present in the Gencode reference annotation.(Table 1; Figure 1a) Despite the large reduction in number of transcripts in the annotation, we see an improvement in transcript quantification rates when using our *de novo* transcriptomes, likely indicating that out constructed transcripts are not an artifact of the construction algorithm. However, we do observe a slight reduction in mapping rates for neuronal derived tissues most likely due to the large number of distinct cell types present in neuronal tissues. Many studies have shown that there are transcriptional isoforms that are highly cell type-specific, so due to the increased number of cell types in neuronal tissue, these transcripts may seem particularly lowly expressed in bulk RNA-seq and thus removed from our dataset.
      In our ocular transcriptomes, we see that novel isoforms are largely subtissue-specific. This matches previously reported findings about the tissue specificity of rare exons and first and last exons. We also find that on average novel isoforms represent about 25% of their parent gene’s total expression, but in cornea tissues the average is significantly higher at about 65%. This is likely due to relatively small size of our cornea transcriptomes. Because we have a relatively lower number of samples and studies for our cornea samples, it is likely that some of the lower expressed novel isoforms were removed during the filtering step of transcriptome construction. We find that within the ocular transcriptome, novel isoforms do not lead to a change in the encoded protein product and lie within the untranslated region of the gene. It is difficult to identify the exact role of these novel exons, but others have shown that the untranslated regions of transcripts contain multiple types of regulatory elements.
      We want to make our transcriptomes easily accessible to other researchers, so we designed a webapp to visualize our transcriptomes and access tissue-specific annotation files. We wanted to provide as much information to the user so that they can draw their own conclusions about the significance of potential novel exon, and so we provide the gene model with novel exon, coding and noncoding regions marked, along with the PIU for each transcript constructed within a gene. We also provide the fraction of samples within a given subtissue type a sample was detected in, to provide a further level of evidence to the validity of constructed transcripts.
      In the future, we aim to provide a more refined ocular transcriptome by using long read RNA-seq data from ocular tissues. Long read RNA-seq captures entire RNA transcripts and thus removes ambiguity due to short read noise or algorithmic construction and will provide a more accurate characterization of the pan eye transcriptome.

::: {custom-style="anchor"}
7-close
:::

## Supplemental Figures

![](plots/dag.png)

::: {custom-style="CustomCaption"}
Supplemental Figure 1. Directed Acyclic Graph of *de novo* transcriptome construction pipeline. 
:::


```{r fig.height=7, fig.width=4}
load('/Volumes/data/ocular_transcriptomes_paper/clean_data/rdata/sup_fig_data.Rdata')
base<- ggplot(data = raw_eye_det)+
    geom_boxplot(aes(x=ano_type, y=frac_det)) + 
    facet_wrap(~subtissue)  + 
    ggtitle('Fraction of total samples transcripts are constructed in, no filtering') +
    theme_minimal()
gfc_filt <- ggplot(data = gfc_eye_det)+
    geom_boxplot(aes(x=ano_type, y=frac_det)) + 
    facet_wrap(~subtissue)  + 
    ggtitle('Fraction of total samples transcripts are constructed in, filtering by study') +
    theme_minimal()
final <- ggplot(data = final_eye_det)+
    geom_boxplot(aes(x=ano_type, y=frac_det)) + 
    facet_wrap(~subtissue)  + 
    ggtitle('Fraction of total samples transcripts are constructed in, final') +
    theme_minimal()
ggarrange(base, gfc_filt, final, ncol=1, labels = "AUTO") 
```

::: {custom-style="CustomCaption"}
Supplemental Figure 2. Detection of samples during stages of construction pipeline
:::



# References