---
title: '*De novo* transcriptomes built from hundreds of eye tissues reveal hundreds of novel gene isoforms'
author: Vinay S Swamy, Temesgen D Fufa, Robert B Hufnagel, David M McGaughey
bibliography: ocular_txome_citations.bib
csl: nucleic-acids-research.csl
output:
  redoc::redoc:
    reference_docx: reference_doc_v1.docx
    keep_md: yes
  html_document:
    df_print: paged
  pdf_document:
    df_print: kable
---

```{r setup, include=FALSE,echo=FALSE, cache = FALSE}
require("knitr")
## setting working directory
#root = '/Users/mcgaugheyd/git/ocular_transcriptomes_paper/'
#opts_knit$set(root.dir = "~/git/ocular_transcriptomes_paper/")
#opts_knit$set(root.dir = root)
#redoc cmd : dedoc(docx, wrap = 9999999, overwrite = T, verbose = T, track_changes = "criticmarkup")
opts_knit$set(root.dir = '/data/swamyvs/ocular_transcriptomes_paper/')
opts_chunk$set(echo = F, message = F, warning = F, include = F)
```

```{r prep, include=FALSE,echo=FALSE}
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(ComplexUpset)
library(ggrepel)
library(ggpubr)
library(patchwork)
library(knitr)
library(grid)
library(png)
library(glue)
sample_table <- read_tsv('sampleTableFullV3.tsv') %>% filter(subtissue != 'synth')
load('clean_data/rdata/tissue_to_colors.Rdata')
load('clean_data/rdata/core_tight.Rdata')
tissue_color_mapping_df <- bind_rows(tissue_color_mapping_df,  tibble(body_location=c('Brain(avg)', 'Body(avg)'), color=c('orange','yellow')))
load('clean_data/rdata/transcriptome_pipeline_stats.Rdata')
subtissue_to_bodyloc <- sample_table %>% select(subtissue, body_location) %>% distinct
core_tight <- core_tight %>% as_tibble %>% select(sample=sample_accession, study=study_accession) %>% distinct
load('clean_data/rdata/paper_numbers.Rdata')
load('clean_data/rdata/longread_summary.Rdata')
load('clean_data/rdata/buildResultsSummary.Rdata')
load('clean_data/rdata/novel_isoforms.Rdata')
load('clean_data/rdata/novel_loci_fig.Rdata')
load('clean_data/rdata/vep_results.Rdata')

color_list<- novel_transcripts_per_tissue$color
names(color_list) <- novel_transcripts_per_tissue$body_location
eye_tissues <-  c("RPE_Fetal.Tissue",  "Retina_Adult.Tissue",  "RPE_Adult.Tissue", "Cornea_Adult.Tissue", "Cornea_Fetal.Tissue", "Retina_Fetal.Tissue")
subtissues <- unique(sample_table$subtissue)
body_tissues <- sample_table %>% filter(!subtissue%in% eye_tissues) %>% pull(subtissue) %>% unique

```

## Introduction

       The transcriptome is defined as the set of distinct RNA transcripts expressed in a population of identical cells. During transcription several RNA processing steps modify immature RNA and drive the formation of multiple, distinct isoforms for most genes. For example{++ , ++} the human Gencode release 28 contains 97,713 protein coding transcripts across 20,306 genes. RNA processing contains a variety of biological mechanisms and includes alternative promoter usage, alternative splicing, RNA editing, and alternative polyadenylation. The full biological impact of gene isoforms has not been fully elucidated, but multiple studies have shown that gene isoforms can have distinct and critical functions in biological processes like development[-@dykes_hic2_2018], cell differentiation[-@trapnell_transcript_2010], and cell migration[-@mitra_splicing_2020]. Alternative usage of isoforms has also been implicated in multiple diseases including cancer[-@vitting-seerup_landscape_2017], cardiovascular diseases[-@noauthor_titin_nodate], Alzheimer's Disease[-@mills_rna-seq_2013] and diabetic retinopathy[-@perrin_diabetic_2005].(didnt read these papers so need to double check them)

Some of the first methods using RNA-seq to find novel gene isoforms focused on identifying novel exon-exon junctions and novel exon boundaries based on RNA-seq coverage.[-@nagalakshmi_transcriptional_2008] More recently, several groups have developed specialized tools to use RNA-seq to reconstruct the whole transcriptome of a biological samples, dubbed *de novo* transcriptome construction [-@haas_novo_2013],(2),[-@pertea_stringtie_2015].
       *De novo* transcriptome construction uses short (\~100bp) RNA-seq reads to reconstruct full-length mRNA transcripts. However, a large number of samples are necessary to overcome the noise and short read lengths e of this type of data. Because of the increasingly inexpensive sequencing, data sets of the necessary size are now available. For example, the most comprehensive *de novo* transcriptome project to date has been CHESS, which used the GTEx dataset to construct *de novo* transcriptomes in over 9000 RNA-seq samples from 49 distinct location of the body to create a comprehensive annotation of mRNA transcripts across the human body. (10),[-@gtex_consortium_genetic_2017],[-@pertea_chess_2018] However, as the GTEx dataset lacks any ocular tissues, the CHESS database is an incomplete annotation of the human transcriptome.

We apply a Stringtie-based approach to construct a pan eye transcriptome using a previously published data set containing `r toString(NUM_EYE_SAMPLES)` ocular tissue samples compiled from mining publicly available sequencing data [@swamy_eye_2019]. We use this pipeline to build transcriptomes in three major ocular subtissues: The cornea, retina, and the RPE, using RNA-seq data from both adult and fetal tissues to create a high-quality pan-eye transcriptome. In addition to our ocular samples, we used a subset of the GTEx dataset to construct transcriptomes for 49 other locations across the body to facilitate comparisons between transcriptomes across the body.

       Despite the increasing number of tools for transcriptome construction there has been no gold standard with which to evaluate precision and sensitivity of de novo transcriptome construction on real (not simulated) biological data. Long read sequencing technologies provide a potential solution to this problem as long read sequencing can capture full length transcripts and thus can be used to identify a fuller range of gene isoforms. While long reads have historically been considered inaccurate, the new PacBio Sequel II system sequences long reads as accurately as short read based sequencing [-@wenger_accurate_2019].

We propose that long read based transcriptomes can serve as a ground truth for evaluating short-read base transcriptomes, and in this study we use PacBio long read RNA sequencing to inform the construction of short read transcriptomes. We generated PacBio long read RNA seq data from a stem cell derived retinal pigmented epithelium (RPE) cell line along with matched Illumina short read RNA-seq. Using the two sources of RNA-seq data we design a rigorous *de novo* transcriptome pipeline that maximizes the {++ concordance ++} between short and long read *de novo* transcriptomes. {++   ++} We use our gold-standard informed pan eye *de novo* transcriptome to reveal hundreds of novel gene isoforms as well as several novel {== transcribed genomic loci  ==}{>> Want to avoid using the word gene <<}in the eye and analyze their potential impact on ocular biology and disease. We provide our *de novo* transcriptomes as a resource to other researchers through an R package{++ . ++}

## Methods


![fig](../clean_data/plots/dntx_flowchart.png){width="6.5in" height="4.826955380577428in"}

::: {custom-style="CustomCaption"}
{==  fig  ==}{>> (TYPOS in figure, also we use illumina short-read data as input, cruddy quality of right top corner) <<} Figure 0. DNTX workflow[]{.paragraph-insertion author="McGaughey, David (NIH/NEI) [E]" date="2020-06-12T10:05:00Z"}
:::
{== {++ Methods ++} ==}{>> Unbold header in ref doc, reduce font size <<}

## Generation of PacBio long read RNA sequencing data and Illumina short read RNA sequencing


Human induced pluripotent stem cells (iPSCs) were differentiated in mature RPE following the culturing protocol in {== {++ first author et al. ++}  ==}{>> Add first author <<}[-@blenkinsop_human_2015]. RNA was isolated from mature RPE 40 days post differentiation and used for \<ilumina library prep\> and \<pacbio library prep\>

## Code availability and software versions.

       To improve reproducibility, we wrote all code used to generate both the data and figures for this paper as Snakemake pipelines.[-@koster_snakemakescalable_2012] All code (and versions) used for this project is publicly available in the following github repositories: <https://github.com/vinay-swamy/ocular_transcriptomes_pipeline> (main pipeline),

<https://github.com/vinay-swamy/ocular_transcriptomes_longread_analysis> (long read analysis pipeline), <https://github.com/vinay-swamy/ocular_transcriptomes_paper> (figures and tables for this paper), <https://github.com/vinay-swamy/ocular_transcriptomes_shiny> ({==  webapp) . ==}{>> $
$move to David's gh ¶ <<}

WE ALSO NEED TO GIVE A GIT TAG FOR EACH TO INDICATE THE "RELEASE" VERSION WITH THIS PAPER (ALSO DEPOSIT IN ZENODO).

## Computational Analyses


All computational analyses performed in this project are run using multiple Snakemake workflows. Each Snakefile contains the exact parameters for all tools and scripts used in each analysis. All Snakefiles are included as supplementary data.(supplementary data files 1-4)

## Analysis of Long Read Data

      Briefly, PacBio hifi reads were processed into full length, non-chimeric (FLNC) read using the Pacbio SMRT link software {== ([VERSION]. We next adapted the ENCODE Long Read analysis pipeline:  Transcripts were aligned to the human genome using the minimap2[-@li_minimap2_2018]  ==}{>> Have to email nisc about this <<}{==  ==}{>> What adaptions??? <<} . Aligned long reads were processed with TranscriptClean[-@wyman_transcriptclean_2019] to remove errors due to sequencing. The long read transcriptome was generated with TALON [-@wyman_technology-agnostic_2020]

## Analysis of short read RPE data

      We aligned each sample to the Gencode release 28 hg38 assembly using the genomic aligner STAR and sorted the resulting BAM files using samtools sort.[-@frankish_gencode_2019],[-@dobin_star_2013],[-@li_sequence_2009]. For each sorted BAM file, we constructed a per sample base transcriptome using StringTie with the Gencode V28 comprehensive annotation as a guiding annotation (17),(10). All sample transcriptomes were merged with the long read transcriptome using gffcompare[-@pertea_gff_2020]. We define the metric construction accuracy, used to evaluate short read transcriptome construction as the following:

$Construction\ Accuracy = \frac {short\ read \ transcriptome\ \cap \ long\ read\ transcriptome} {short\ read \ transcriptome}$

## Construction of tissue specific transcriptomes.


       We used studies with healthy, unperturbed RNA-seq GTEX project samples from `r toString(NUM_BODY_TISSUES + NUM_EYE_TISSUES)` distinct subtissue regions of the body, downloaded and performed quality control the pertinent sequencing data from the sequence read archive (SRA) using methods from our previous work[-@swamy_eye_2019]. We constructed a transcriptome for each sample, and merged samples together to create `r toString(NUM_BODY_TISSUES + NUM_EYE_TISSUES)` subtissue specific transcriptomes. For each tissue-specific transcriptome, we removed transcripts that had an average expression less than 1 Transcripts Per Million (TPM). All tissue specific transcriptomes were merged to form a single unified GTF annotation file to ensure transcript identifiers were the same across tissues. We merged all ocular tissue transcriptomes to generate a separate pan-eye transcriptome.

## Tissue specific transcriptome quantification


       For each resulting tissue specific transcriptome, we extracted transcript sequences using the tool gffread(20) and used these sequences to build a tissue-specific quantification index using the index mode of the alignment free quantification tool Salmon.[-@patro_salmon_2017] For each sample, we quantified transcript expression using the quant mode of salmon, using a samples respective tissue specific quantification index. We similarly quantified all ocular samples using the pan-eye transcriptome and the Gencode v28 reference transcriptome.

## Downstream analysis of novel transcripts

Analysis of novel exons was done using a custom Rscript {==  "annotate\_and\_make\_tissue\_gtfs.R"  ==}{>> This step is part of a larger script might consider separating into its own script. <<}. To predict the clinical significance of clinvar variants of uncertain significance, we selected variants from the ClinVar {== database (which version/date?)  ==}{>> Add clinvar version/date <<}labeled as "Uncertain Significance" and used Ensembl's Variant Effect Predictor tool to predict the consequence of known mutations that fell within novel transcriptome regions we identified.

## Identification of protein coding novel transcripts.

       We identified protein coding transcripts in our unified transcriptome using the TransDecoder suite (9). We extracted transcript sequences using the util script gtf\_genome\_to\_cdna\_fasta.pl and used TransDecoder to find a single best open reading frame from each transcript. We used the agat\_sp\_add\_start\_stop.pl scripts from the AGAT(cite) tool to identify start and stop codons for each ORF. Transcripts with no detectable ORF or missing a start or stop codon were labelled as noncoding.

## Prediction of novel loci function

Protein coding novel loci were compared to the uniprot protein sequence data base using blastp[-@altschul_basic_1990]. Blastp results were integrated with hmmer, which identified protein families and domains associated with each novel loci.

## Computing Resources
-------------------

       All computation was performed on the National Institutes of Health high performance compute system Biowulf (hpc.nih.gov).

## Figures and Tables

       All statistical analyses, figures and tables were generated using the R programming language.[-@r_core_team_r_2019] A full list of packages and versions can be found in supplementary file session\_info.txt

# Results

## Long Read Pacbio RNA sequencing guides *de novo* transcriptome construction

In order to determine the accuracy of short read transcriptome construction, we first generated PacBio long read RNA-seq data and Illumina short read RNA-seq data from a stem cell derived RPE cell line. These cell lines were cultured using a highly optimized protocol, and thus should have minimal biological variation.\<cite\> We used this sequencing data to construct a long read transcriptome and a short read transcriptome. In our long read transcriptome we find `r toString(NUM_TOTAL_LR_TX)` distinct transcripts, and in our short read transcriptome `r toString(NUM_TOTAL_SR_TX)` distinct transcripts

```{r longread_results_prep}
plot_violin <- function(ldf, origin){
  ldf <- ldf %>% filter(`Transcript Origin` == origin)
  sum_counts <- ldf%>% 
    mutate(intersection_case = gsub('stringtie', 'StringTie', intersection_case),
           intersection_case = gsub('pacbio', 'PacBio', intersection_case),
           intersection_case = gsub('StringTie-PacBio', 'PacBio &\nStringTie', intersection_case)) %>% group_by(intersection_case, `Transcript Origin`) %>% count()
  violin_plot <- ldf %>% 
    mutate(intersection_case = gsub('stringtie', 'StringTie', intersection_case),
           intersection_case = gsub('pacbio', 'PacBio', intersection_case),
           intersection_case = gsub('StringTie-PacBio', 'PacBio &\nStringTie', intersection_case)) %>% 
    ggplot() +
    scale_y_continuous(breaks = c(2000,4000,6000, 8000, 10000)) +
    geom_text(data = sum_counts, aes(x=intersection_case, y = 11000, label=n), stat= 'identity') +
    geom_violin(aes(x=intersection_case,y=length)) + geom_hline(yintercept = 2000, color='red') +
    cowplot::theme_cowplot() + ylab(glue('{origin}\nTranscript Length (bp)')) + xlab('') + coord_flip(ylim = c(0,11500)) 
  return(violin_plot)
}
violin_plot_novel <- plot_violin( length_df_plotting, 'Novel')
violin_plot_gencode <- plot_violin(length_df_plotting, 'Gencode')
tpm_exp_plot <-  ggplot(data = tpm_df_plotting %>% ungroup, aes(x=co, y=acc, color=`Length Interval`)) + 
  geom_point() + 
  geom_line()+
  #xlim(c(-3,26))+
  #ylim(c(0,1.3))+
  #geom_label_repel(data = labdf_plotting%>% ungroup, aes(x=co, y=y, label = total_tx), fill='white', color='black', nudge_x = -1.5, nudge_y = .25) +
  labs(label = 'asf')+
  ylab('Build Accuracy\n(StringTie ∩ PacBio / StringTie)') +
  xlab('TPM threshold')  +
  facet_wrap(~filtering_type) +
  cowplot::theme_cowplot() +
  scale_color_viridis_d()



```
```{r longread_results, fig.height=7, fig.width=8, include=T, echo = T}
violin_plot_novel/violin_plot_gencode/tpm_exp_plot +plot_annotation(tag_levels = 'A') + plot_layout(heights = c(1,1,1.7))
```


::: {custom-style="CustomCaption"}
Figure 1. Transcript length is substantially larger in the long-read based approach. A,B) Intersection of novel and annotated transcript lengths between Pacbio (long read) and Stringtie (short read) transcriptomes. The total number of constructed transcripts is given in the text to the right of the violin plot. C) Short read construction accuracy stratified by transcript length at different TPM based transcript exclusion thresholds.
:::

    In our initial comparison between short and and long read transcriptomes, we see a low transcriptome construction accuracy (see Methods) of `r toString(NUM_BASE_TXOME_CONST_ACC)`. When examining the transcript lengths of each build we see that the two methods show very different transcript length distributions for both novel and previously annotated transcripts, with the short read build comprised mostly of smaller transcripts (Fig 1A). As the PacBio data was generated using two libraries for {== 2000 bp and \>3000 bp  ==}{>> $$\text{PUT\ IN\ \ MEDIAN\ AND\ sMEAN}$$ <<}(need to double check this), we expected to see enrichment for longer transcripts in the pacbio.(sup fig 1) To assess accuracy relative to transcript length, we group transcripts by length in 1000 bp intervals, and compare accuracy between each group. We found the accuracy significantly improves for transcripts longer than 2000 bp. The construction accuracy is `r toString(NUM_TXOME_CONST_ACC_2000_ABOVE)` and `r toString(NUM_TXOME_CONST_ACC_2000_BELOW)` for transcripts above and below 2000 bp, respectively.\
    We experimented with various methods to remove spurious transcripts to improve construction accuracy. We first removed transcripts that were not expressed at 1 TPM in at least one sample as outlined in the stringtie's recommended protocol.[-@pertea_transcript-level_2016] This improved construction accuracy to `r toString(NUM_TXOME_MERGE_ACC_2000_ABOVE)` for transcripts longer than 2000bp and `r toString(NUM_TXOME_MERGE_ACC_2000_BELOW)` for transcripts shorter than 2000bp. As this accuracy was still fairly low, we tried other filtering schemes (data not shown?). We found the simplest approach with high performance is to retain transcripts that had an average of 1 TPM expression across all short read samples. This achieved a build accuracy of `r toString(NUM_FILT_TXOME_ACC_2000_ABOVE)` and retained `r toString(NUM_TXOME_FILTER_SIZE)` transcripts. In our downstream pipeline we keep transcripts that have an average TPM of at least one across all samples of the same subtissue type.

## A rigorous analysis pipeline finds thousands of novel gene isoforms

```{r overall_stats_table}
df <- sample_table %>% 
  left_join(core_tight) %>% 
  filter(!body_location %in% c('Body', 'Brain', 'ESC_Stem.Cell.Line', 'Lens_Stem.Cell.Line')) %>%
  group_by(subtissue) %>%
  summarise(`Samples` = n(), `Studies` = length(unique(study))) %>% 
  arrange(desc(`Studies`)) %>% 
  inner_join(tx_counts %>% select(subtissue, `Transcriptome Count`=filtered)) %>% 
  mutate(subtissue=gsub('_|\\.',' ', subtissue)) %>% 
  mutate(subtissue = gsub(' Tissue', '', subtissue)) %>% 
  separate(subtissue, into  = c('Tissue' , 'Source'), sep = ' ')
kable(df)
```

::: {custom-style="CustomCaption"}
Table 1. Ocular sample dataset overview and transcriptome size. Transcriptome count is defined as the number of unique transcripts expressed in a given tissue type
:::

      We built transcriptomes from `r toString(NUM_EYE_SAMPLES)` published, publicly available ocular tissue RNA-seq samples using an efficient snakemake pipeline. We include both adult and fetal tissue from cornea, retina, and RPE tissues mined from `r toString(NUM_STUDIES)` different studies (Table 1). Our fetal tissues consist of both human fetal tissues and human induced pluripotent stem cell (iPSC) derived tissue, as stem cell derived tissue has been showed to closely resemble fetal tissue. (CITATIONS)

To more accurately determine tissue specificity of novel ocular transcripts, we supplemented our publicly collated normal (non-disease, perturbation) ocular data set with `r toString(NUM_BODY_SAMPLES)` samples across `r toString(NUM_BODY_TISSUES)` body locations from the GTEx project and constructed transcriptomes for each of these body locations.(cite) We refer to each distinct body location as a subtissue here after.
      After initial construction of transcriptomes, we found `r toString( NUM_TOTAL_GENCODE_TX)`  previously annotated transcripts  and ZZZ novel transcripts detected in at least one of our `r toString(NUM_EYE_SAMPLES+NUM_BODY_SAMPLES)` samples. We define novel as any region of the human genome that has not been previously annotated within the Gencode, Ensembl and Refseq annotation databases. After using the filtering methods desribed above, our final transcriptome across all tissues contains  `r toString(NUM_TOTAL_TX_BODY)` distinct transcripts with `r toString(NUM_REF_TX_BODY)`  previously annotated and `r toString(NUM_NOVEL_TX_BODY)` novel transcripts, and includes `r toString(NUM_TOTAL_NOVEL_SEQ/1000000 %>% round(digits=1))` megabases of previously unannotated genomic sequence. [-@frankish_gencode_2019]    ,[-@zerbino_ensembl_2018],[-@oleary_reference_2016].
      We split novel transcripts are into two categories: novel isoforms, which are novel variations of known genes, and novel loci, which are previously unreported, entirely novel regions of transcribed sequence. Novel isoforms are further classified by the novelty of its encoded protein: an isoform with novel open reading frame, a novel isoform with a known ORF, and isoforms with no ORF as noncoding isoforms. The number of distinct ORFs is significantly less than the number of transcripts, with `r toString(NUM_REF_ORF)` previously annotated ORFs and `r toString(NUM_NOVEL_ORF)` novel ORFs across all tissues. Across all tissues there is an average of `r toString(NUM_AVG_NOVEL_ISOFORM_PER_TISSUE)` novel isoforms and `r toString(NUM_AVG_NOVEL_ORF_PER_TISSUE)` novel ORFs.

```{r overall_stats_figure, fig.height=7, fig.width=8}

novel_isoforms_per_tissue <- inner_join(novel_transcripts_per_tissue, novel_orfs_per_tissue) %>% 
  select(-novel_pc_count) %>% 
  gather(key = 'transcript_type', value = 'counts', -body_location, -color) %>% 
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=gsub('Tissue','', body_location_pretty),
         body_location_pretty=gsub('avg','mean', body_location_pretty),
         body_location_pretty=gsub('\\(',' (', body_location_pretty),
         body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
         body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
  mutate(transcript_type = gsub('_',' ', transcript_type),
         transcript_type = gsub(' count', '', transcript_type),
         transcript_type = gsub('nc','NC', transcript_type),
         transcript_type = gsub('orf','ORF', transcript_type),
         transcript_type = gsub('novel','Novel', transcript_type),
         transcript_type = gsub('ref','Ref.', transcript_type))
novel_loci_per_tissue <- novel_loci_per_tissue %>% 
  gather(key = 'transcript_type', value = 'counts', -body_location, -color) %>% 
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
  mutate(transcript_type = gsub('_\\w+','', transcript_type),
         transcript_type = gsub('noncoding' ,'Noncoding', transcript_type),
         transcript_type = gsub('pc','Protein\nCoding', transcript_type)) %>% 
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=gsub('avg','mean', body_location_pretty),
         body_location_pretty=gsub('Tissue','', body_location_pretty),
         body_location_pretty=gsub('\\(',' (', body_location_pretty),
         body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
         body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty))
color_list<- novel_isoforms_per_tissue$color
names(color_list) <- novel_isoforms_per_tissue$body_location_pretty

loci <- ggplot(data = novel_loci_per_tissue) +
  geom_col(aes(x=body_location_pretty, y=counts, fill = body_location_pretty, alpha = transcript_type), position = 'dodge') +
  scale_fill_manual(values = color_list, 'Body\nLocation') +
  scale_alpha_discrete(range=c(.5,1), name = 'Transcript\nType') +
  #ggtitle('Novel Isoforms Contructed Across the Body')+
  ylab('Novel\nTranscript\nCount')+
  xlab('Body Location')+
  cowplot::theme_cowplot() + 
  theme(axis.text.x=element_blank(), panel.background = element_blank())

isotype_per_tissue <- ggplot(data = novel_isoforms_per_tissue) +
  geom_col(aes(x=transcript_type, y=counts, fill = body_location_pretty), position = 'dodge') +
  scale_fill_manual(values = color_list) +
  scale_alpha_discrete(range=c(.5,1)) +
  #ggtitle('Novel Isoforms Contructed Across the Body')+
  ylab('Noncoding Transcript Count')+
  xlab('')+
  cowplot::theme_cowplot() + theme(legend.position = "none") + 
  theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank())+
  facet_wrap(~body_location_pretty, ncol=2)

novel_transcript_types <-  novel_isoform_anno_df %>% filter(label != 'Protein Coding') %>% 
  mutate(group = 'Transcript Type') %>% 
  rename(`Transcript Type` = label) %>% 
  ggplot(data = ., aes(x = group, fill = `Transcript Type`, y=count)) +
  geom_bar(position = 'fill',stat = 'identity') +
  geom_text(aes(label = count), position = position_fill(vjust=.5)) +
  ylab('% of all novel transcripts') +
  cowplot::theme_cowplot() + 
  theme(panel.background = element_blank(), axis.title.x = element_blank())

novel_exon_types <- exon_type_by_transcript_type %>% group_by(nv_type_rc) %>% 
  summarise(count = n()) %>% 
  mutate(group =  'Novel Exon Type') %>% 
  rename(`Novel Exon Type` = nv_type_rc) %>% 
  ggplot(data = ., aes(x = group, fill = `Novel Exon Type`, y=count)) +
  geom_bar(position = 'fill',stat = 'identity') +
  geom_text(aes(label = count), position = position_fill(vjust=.5)) +
  ylab('% of all novel transcripts') +
  cowplot::theme_cowplot() +
  theme(panel.background = element_blank(), axis.title.x = element_blank())

exon_type_tx_plot <- exon_type_by_transcript_type %>% 
  mutate(comp_transcript_type = gsub('Isoform', 'Isoform,\n', comp_transcript_type)) %>% 
  mutate(nv_type_rc = gsub('_',' ', nv_type_rc),
         nv_type_rc = gsub('novel', 'Novel', nv_type_rc)) %>% 
  group_by(comp_transcript_type, nv_type_rc) %>% 
  summarise(count = n()) %>% 
  ggplot(data=., aes(x  = comp_transcript_type,y=count, fill = nv_type_rc)) +
  geom_bar(position = 'fill', stat = 'identity')+
  geom_text(aes(label = count), position = position_fill(vjust=.5)) +
  ylab('Proportion of Transcripts') +
  cowplot::theme_cowplot() +
  theme(axis.text.x=element_text(angle=45, hjust = 1), 
        panel.background = element_blank(), axis.title.x = element_blank()) +
  scale_fill_manual(values = pals::cols25() %>% unname(), name = 'Transcript\nCategory')

layout <- '
AB
AC
'
isotype_per_tissue + loci + exon_type_tx_plot +plot_annotation(tag_levels = 'A') +plot_layout(design = layout, heights = c(1,5), widths = c(1.5,1), guides = 'collect') 

```

::: {custom-style="CustomCaption"}
Figure 2. Overview of Novel Isoforms. A. Number of novel gene isoforms, grouped by transcript type. Brain and body represent an average of 13 and 34 distinct subtissues, respectively B. Novel protein coding and noncoding loci. Novel exon composition of novel isoforms, by isoform type labels indicate number of transcripts.
:::

       Novel isoforms can occur due to an omission of a previously annotated exon, commonly referred as exon skipping or the addition of a previously unannotated exon which we define as a novel exon. We further classify novel exons by the biological process that may be driving their formations: alternative promoter usage driving the addition {== of novel first exons,  ==}{>> ? <<}alternative polyadenylation driving the addition of novel terminal exons, and alternative splicing driving the formation of all novel exons that are not the first or last exon. We further classify alternatively spliced exons into their commonly seen patterns, alternative 5' splice site (A5SS), alternative 3' splice site (A3SS), and retained intron. We note that all three of these mechanisms can lead to exon skipping, so for simplicity we group all novel isoforms resulting from exon skipping together. (considering adding a sup fig with diagrams of each group to make it easier for ppl)

We found that the majority of novel exons with our dataset are novel first and last exons. We see that the majority of A5SS exons lead to novel protein coding isoforms, whereas novel TES/TSS more often lead to noncoding isoforms. Retained introns are most commonly predicted to lead to a novel ORF.

## *de novo* transcriptomes reduce overall transcriptome sizes

```{r map_rate_diff, fig.height=5, fig.width=4}
map_rate_diffs <- all_sample_mapping_rate_difference %>%
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location=case_when(body_location == 'Body' ~ 'Body(mean)',
                                 body_location == 'Brain'~ 'Brain(mean)' ,
                                 TRUE ~ body_location),
         body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=gsub('avg','mean', body_location_pretty),
         body_location_pretty=gsub('Tissue','', body_location_pretty),
         body_location_pretty=gsub('\\(',' (', body_location_pretty),
         body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
         body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty),
         mapping_rate_diff = mapping_rate_diff * -1 ,
         mapping_rate_diff = replace(mapping_rate_diff, mapping_rate_diff <=0, 0),
         pt = 'mapping rate reduction')

size_reduction_df  <-  inner_join(gencode_tx_size, tx_counts) %>%
  mutate(percent_txome_size_decrease = ( all_exp-filtered) / all_exp) %>%
  inner_join(subtissue_to_bodyloc) %>%
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location=case_when(body_location == 'Body' ~ 'Body(mean)',
                                 body_location == 'Brain'~ 'Brain(mean)' ,
                                 TRUE ~ body_location)) %>% 
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=gsub('avg','mean', body_location_pretty),
         body_location_pretty=gsub('Tissue','', body_location_pretty),
         body_location_pretty=gsub('\\(',' (', body_location_pretty),
         body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
         body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
  group_by(body_location_pretty) %>%
  summarise(percent_txome_size_decrease = mean(percent_txome_size_decrease)) %>%
  mutate(pt = '% Base txome removed')

cl<- novel_transcripts_per_tissue$color
names(cl) <- novel_transcripts_per_tissue$body_location
ggplot(map_rate_diffs) +
  geom_boxplot(aes(x = pt, y = mapping_rate_diff, fill = body_location_pretty))+
  geom_col( data = size_reduction_df, aes(x = pt, y=percent_txome_size_decrease, fill = body_location_pretty)) +
  scale_fill_manual(values = color_list, ) +
  theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank()) +
  facet_wrap(~body_location_pretty, ncol = 2) +
  ylab('fraction decrease') +
  theme(legend.position = "none")

```

::: {custom-style="CustomCaption"}
Figure 3. Comparison of Salmon mapping rate decrease vs transcriptome size decrease. Any gain in mapping rate has been set to 0 for {-- simiplicity --}{++ simplicity ++}
:::
      Our transcriptomes removed on average of `r toString(NUM_AVG_TXOME_SIZE_DECREASE*100)` % of a tissue's base transcriptome. We define base transcriptome for a tissues as any transcript in the gencode annotation with non zero TPM in at least one sample of a given tissue type. While this result was consistent with the CHESS project result (refer to figure in CHESS paper?), this was a large reduction in transcriptome size. We wanted to ensure we were not unduly throwing away data. We quantified transcript expression of our samples using Salmon, quantifying each sample twice: once using the full gencode V28 human transcript annotation, and once using its associated tissue specific transcriptome We found that despite the `r toString(NUM_AVG_TXOME_SIZE_DECREASE*100)` % reduction in number of transcripts between the base gencode and denovo transcriptomes, there is only an average decrease of `r toString(NUM_AVG_MAP_RATE_DIF*100)` % in mapping rate, indicating that majority of gene expression data is preserved.

## *de novo* transcriptomes identify of the tissue specific primary isoform expressed in a tissue

    It is useful researchers to know which isoform of a gene is the most relevant to a particular biological system. The APPRIS project identified principal isoforms for humans, based on predicted function and evolutionary conservation, using multiple computational tools. Principal isoforms are considered the most phenotypically relevant isoforms of transcripts. We instead propose a tissue specific primary isoforms, defined as the highest expressed isoforms of a given gene in a given tissue. As primary isoforms are based on biological evidence via RNA-seq, we reason that it is more refined than APPRIS principal isoforms.

```{r, primary_isoforms, fig.width= 6}

primary_isoforms_per_tissue <- primary_isoforms_per_tissue %>%
  mutate(primary_isoform_origin = gsub('dntx','DNTx', primary_isoform_origin),
         primary_isoform_origin = gsub('gencode', 'GENCODE', primary_isoform_origin),
         primary_isoform_origin = gsub('appris', 'APPRIS', primary_isoform_origin)) %>% 
  mutate(primary_isoform_origin = factor(primary_isoform_origin, levels = c('DNTx', 'GENCODE', 'APPRIS'))) %>% 
  inner_join(subtissue_to_bodyloc) %>%
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>%
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=case_when(body_location == 'Body' ~ 'Body(mean)',
                                        body_location == 'Brain'~ 'Brain(mean)' ,
                                        TRUE ~ body_location)) %>% 
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),
         body_location_pretty=gsub('avg','mean', body_location_pretty),
         body_location_pretty=gsub('Tissue','', body_location_pretty),
         body_location_pretty=gsub('\\(',' (', body_location_pretty),
         body_location_pretty=gsub('Adult','(Adult)', body_location_pretty),
         body_location_pretty=gsub('Fetal','(Fetal)', body_location_pretty)) %>% 
  group_by(body_location_pretty, primary_isoform_origin) %>%
  summarise(count = mean(count)) %>% 
  arrange()
ggplot(data = primary_isoforms_per_tissue, aes(x = body_location_pretty, y=count, fill = primary_isoform_origin))+
  geom_bar(position = 'fill', stat = 'identity')+
  geom_text(aes(label = as.integer(count) ), position = position_fill(vjust=.5)) +
  cowplot::theme_cowplot() +
  theme(axis.text.x=element_text(angle=45, hjust = 1), panel.background = element_blank()) +
  scale_fill_brewer(palette = 'Set1', name = 'Origin') + ylab('Ratio') + xlab('')

```

::: {custom-style="CustomCaption"}
Figure 4. APPPRIS principal isoforms within the *de novo* transcriptome
:::

   {== We found that `r toString(NUM_APPRIS_RECOVERY_RATE)` of the APPRIS principal isoforms are also identified as the primary isoform in at least one tissue type. Within each tissue type, on average `r toString(NUM_AVG_APPRIS)` % of are an appris primary isoform, `r toString(NUM_AVG_NONAPPRIS_GENCODE)` of transcripts were a previously annotated gencode transcript, `r toString(NUM_AVG_NOVEL)`

\% of principal isoforms are novel isoforms. ==}{>> need to fix these numbers <<}

## Novel Isoforms in Ocular tissues

```{r, novel_isoforms_ocular_tissues, fig.width=10, fig.height=5}
source('scripts/R_raincloud.R')
plot_list <- novel_eye_tx_by_tissue[!names(novel_eye_tx_by_tissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line') ]
all_tx <- tibble(transcript_id = reduce(plot_list, union))
plot_df <- bind_cols(all_tx, lapply(plot_list, function(x) all_tx$transcript_id%in% x))
FIG_FONT_SIZE=12

colnames(plot_df) <- gsub('_|\\.', ' ', colnames(plot_df) )                             
colnames(plot_df) <- gsub('Adult', '(Adult)', colnames(plot_df) )
colnames(plot_df) <- gsub('Fetal', '(Fetal)', colnames(plot_df) )
colnames(plot_df) <- gsub('Tissue', '', colnames(plot_df) )

us <- ComplexUpset::upset(data = plot_df, 
                          intersect = colnames(plot_df)[-1], 
                          min_size= 500, 
                          set_sizes = F,
                          #height_ratio = .8, 
                          #width_ratio = .1,
                          wrap = T,
                          stripes = c('white','white'),
                          base_annotations = list(
                            'Intersection size' = intersection_size(text_colors = c(on_background='red', on_bar='red'))
                          ),
                          themes=upset_modify_themes(
                            list(
                              "intersections_matrix"= theme(text = element_text(size=FIG_FONT_SIZE) ),
                              "Intersection size"= theme(text = element_text(size=FIG_FONT_SIZE),
                                                         panel.grid.major = element_blank(), 
                                                         panel.grid.minor = element_blank()),
                              "overall_sizes"= theme(axis.text.x=element_text(angle=45),text = element_text(size=FIG_FONT_SIZE) ),
                              "default"= theme(text = element_text(size=FIG_FONT_SIZE)) 
                            )
                          )
                          
)

plotting_piu_df <- piu_df %>% mutate(new_stage=ifelse(stage == 'adult', 'Adult', 'Fetal'), 
                                     subtissue=paste0(tissue, '_', new_stage, '.Tissue')) %>% 
  mutate(subtissue=gsub('_|\\.', ' ', subtissue),
         subtissue=gsub('avg','mean', subtissue),
         subtissue=gsub('Tissue','', subtissue),
         subtissue=gsub('\\(',' (', subtissue),
         subtissue=gsub('Adult','(Adult)', subtissue),
         subtissue=gsub('Fetal','(Fetal)', subtissue))

piu <- ggplot(data=plotting_piu_df, aes(x=stage, y=piu, fill=subtissue, color=subtissue)) + 
  geom_flat_violin(position = position_nudge(x = .2, y = 0),adjust = 2)+
  geom_point(position = position_jitter(width = .15), size = .25, alpha=.2) + 
  geom_boxplot(outlier.shape = NA, alpha = 1, width = .1, colour = "BLACK")+ 
  scale_fill_manual(values = color_list) +
  scale_color_manual(values = color_list) +
  ylab('piu')+xlab('tissue')+coord_flip()+ 
  cowplot::theme_cowplot()+ facet_wrap(~tissue, nrow = 3) +
  theme(legend.position = "none")

us + piu  + plot_layout(widths = c(2,1)) + plot_annotation(tag_levels  = 'A') & theme(plot.tag = element_text(size= FIG_FONT_SIZE), text = element_text(size = FIG_FONT_SIZE))

```

::: {custom-style="CustomCaption"}
Figure 5. Analysis of novel isoform characteristics A. Set intersection of novel isoforms in ocular transcriptomes. B Boxplots of PIU overlaid over PIU data points with estimated distribution of data set above each boxplot
:::

      Next, we analyzed the novel isoforms within our ocular transcriptomes. We compared the overlap in constructed novel isoforms across ocular tissues and found that XXX % of novel isoforms are specific to a singular ocular subtissue (Fig 3A{++ ) ++}. For each novel isoform we then calculated percent isoform usage (PIU), or the fraction of total gene expression a transcript contributed to its parent gene. We found that on average novel isoforms contribute to 15% of their parent gene's expression, although we see multiple novel isoforms that are the primary isoform for their parent gene.

```{r}
# ## Novel Loci in the ocular transcriptome
# 
# We determined the longest ORF’s for each novel loci, and for each transcript with a valid ORF, predicted the function the encoded protein using hmmer searching within the pfam databse, and blastp searching within the swissprot database. We chose to hmmer results that had a sequence match e-value < 1e-4, and blastp results with e-value < 1e-15 (Sup Table). We found XXX distinct potentially novel coding loci from blast and hmmer combined. We highlight one detected in fetal RPE
# library(patchwork)
# 
# cov <- ggplot(cov_df, aes(x=coord,y=cov, group = sample))+
#   geom_line( alpha = .3)+
#   geom_area(alpha = .3)+
#   #ggtitle(paste(txid, s_st))+
#   theme(axis.title.x = element_blank(), panel.background = element_blank(), 
#         axis.text.x = element_blank(), axis.ticks.x = element_blank())
# pt <- ifelse(gene_model_df$strand[1] == '+',62, 60)
# model <- ggplot(gene_model_df)+
#   geom_rect(aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax))+
#   geom_rect(xmin=min(gene_model_df$xmin), xmax=max(gene_model_df$xmax), ymin=-.1, ymax=.1)+
#   geom_point(data=point_df, aes(x = x, y=y), shape= pt, size = 5)+
#   xlab('genome') +
#   ylab(element_blank())+
#   theme(panel.background = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank())
# layout <- '
# AA
# AA
# AA
# AA
# AA
# AA
# BB
# CC
# CC
# '
# mss <- as_ggplot(rasterGrob(readPNG('clean_data/plots/msa_ex.png'), interpolate = T))
# cov/model/mss + plot_layout(design = layout) +plot_annotation(tag_levels = 'A')
# 
# ::: {custom-style="CustomCaption"}
# Figure 6. DNTX000dddd, potentially protein coding novel isoform in Fetal RPE. A) RNA-seq coverage of exons across 49 samples. B transcript model of DNTX000ddd. thick regions represtn coding region. C. Multiple sequence alignment of DNTX000ddd and closely related proteins.
# :::
# 
# Placeholder figure, still trying to find the right transcript to look at here.
```

## *de novo* transcriptome allow for a more precise variant prioritization.

The prediction of variant's biological impact is a fundamental step in variant prioritization, especially in the diagnosis of genetic diseases or in interpreting GWAS results. We use Ensembl's Variant Effect Predictor combined with our tissue transcriptomes to generate a tissue specific prediction of variant impact for variants of unknown significance (VUS) the clinvar database. We grouped different predicted consequences into a High, moderate and low impacts, as defined by the Variant Effect Predictor, and provide a full mapping of biological consequence to impact in our supplementary data.

```{r vep, fig.width=16, fig.height=12}
library(tidyverse)
cl <- c('Low' = 'lightblue', 'Moderate' = 'lightgreen', 'High' = 'red')
shp <- c('dntx'=16, 'gencode'=15)
k <- vep_plot_df %>% 
  filter(!as.character(subtissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
  left_join(subtissue_to_bodyloc) %>% mutate(body_location = replace_na(body_location, 'gencode') %>% factor(levels = c('gencode', eye_tissues, 'Body', 'Brain'))) %>% group_by(body_location, allele_id, build) %>% 
  summarise(`avg log2(TPM + 1)` = mean(`log2(TPM + 1)`), avg_max_impact =mean(max_impact) )
ggplot(k, aes(x =allele_id, y=body_location))+
  geom_point(aes(size = `avg log2(TPM + 1)`, color = `avg_max_impact`, shape=build))+
  scale_color_gradientn(colors =RColorBrewer::brewer.pal(n=6,name = 'YlOrRd'))+
  #scale_color_gradientn(colors=rev(viridisLite::viridis(10)))+
  scale_shape_manual(values = shp)+
  geom_hline(yintercept = 1.5)+
  ylab('Subtissue')+
  xlab('ClinVar Allele ID')+
  ggtitle('Tissue Specific Variant Prioritization') +
  cowplot::theme_cowplot() +
  theme(axis.text.x.bottom = element_text(angle = 45, hjust = 1), panel.background = element_blank())

```

::: {custom-style="CustomCaption"}
{==  {==  Figure 7. Tissue specific variant prioritization of variants of unknown significance {++   ++} (VUS) previously associated with ocular disease. Size corresponds to average TPM and color to impact. A full mapping of variant descriptors to impact can be found in supplemental data ==}{>> Clean up formatting (underscores and periods). ¶ <<}.  ==}{>> Color scheme should have three levels \_1,2,3 -- so make this a factor -- easier to evaluate. <<}
:::

We compared out tissue specific variant effect predictions to variant effect predictions using the gencode annotation. We found that out of `r toString(NUM_TOTAL_VUS)` total variants examined, the highest priority of `r toString(NUM_IMPACT_DECREASE)` decreased in at least one tissue, and `r toString(NUM_IMPACT_INCREASE)` increased in at least one tissue. We see a larger decrease in predicted effects as the transcript which previously lead to an effect is not expressed in a respective tissue.[]{.paragraph-insertion author="McGaughey, David (NIH/NEI) [E]" date="2020-06-12T14:45:00Z"}

We examined ClinVar variants of unknown significance (VUS) that were associated with an ocular disease phenotype to see if our tissue specific variant predictions could provide new insight on. We found that for ClinVar variant 280266, which was previously associated retinitis pigmentosa is only considered high impact in retinal but not in other tissue, and so this variant more may be more likely to be damaging. Variant 843125, which is associated with retinal pigment dystrophy has a high impact in other tissues, but a low impact in all ocular tissue and so likely not disease causing.

## A companion visualization tool enables easy use of *de novo* transcriptomes

```{r viz, fig.height=6}
piu <- as_ggplot(rasterGrob(readPNG('clean_data/plots/piu_shiny_app.png'), interpolate = T))
frac_det <- as_ggplot(rasterGrob(readPNG('clean_data/plots/frac_sample_det.png'), interpolate = T))
gene_body <- as_ggplot(rasterGrob(readPNG('clean_data/plots/gene_bodies_shiny_app.png'), interpolate = T))
layout <- '
AABB
CCCC
'
piu + frac_det + gene_body + plot_layout(design = layout)+ plot_annotation(tag_levels = 'A')

```

::: {custom-style="CustomCaption"}
Figure 8. Screenshots from dynamic *de novo* transcriptome visualization tool. A). PIU bar plot for selected gene and tissue. B). Exon level diagram of transcript body Thicklines represent coding region of transcript. novel exons colored in red. Tooltip contains genomic location and phylop score C) Bargraph of fraction of samples within dataset each transcript was consructed in by tissue.
:::

       To make our results easily accessible, we designed a web-app for visualizing and accessing our *de novo* transcriptomes. Users start by selecting a gene or searching for a gene by genomic location, and can choose up to 5 tissues to visualize transcript expression in. For each tissue we show the PIU for each transcript associated with a gene. We show the exon-intron structure of each transcript and mousing over exons show genomic location overlapping SNPs, and phylogentic conservation score. We additionally show a barplot of the fraction of samples in each tissue each transcript was constructed in. Users can {== also download the *de novo* transcriptomes for selected tissues in GTF and fasta format ==}{>> Add something about adding dntx to eyeintegration <<}.

## A easy-to-use pipeline to enable future research

{==  Finally, we package all tools used for our transcriptome pipeline within a portable docker container with a stand-alone run script. This pipeline allows other researchers to run their own samples, and generate figures and annotations similar to what is shown here ==}{>> . (havent worked it out entirely, but snakemake can automatically use a singulaity container, so if I provide a singularity image anyone can use it) ¶ <<}

{== Discussion ==}{>> DISCUSSION IS TOO LONG CUT DOWN <<} {#discussion-is-too-long-cut-downdiscussion}
# Discussion

      Motivated by the lack of a comprehensive pan eye transcriptome we created the first comprehensive set of ocular transcriptomes. By using long read RNA-sequencing data to calibrate our short-read construction pipeline, we increase our confidence that we are creating real, biologically relevant transcriptomes. We found that concordance between long and short read based transcriptome is directly related to transcript length and transcript expression across many tissues.

We saw a clear inability within this PacBio data set to accurately detect transcripts shorter than 2000Bp for both previously annotated and novel transcripts. As many of the transcripts constructed using short reads are below this threshold, long read sequencing data enriched for smaller transcript sizes would provide greater insight in future studies. Even when accounting for transcript length, {== we achieved a construction accuracy of `r toString(NUM_TXOME_CONST_ACC_2000_ABOVE)`.  ==}{>> Didn't we get to 0.7?(wrong number) <<}This was surprisingly low. It is difficult to say whether this low concordance is due to a high degree of transcriptional noise or errors in the short read transcriptome construction.

We see many transcripts in the long read data that were not corroborated in our short-read based transcriptomes. An approach that uses many long-read based transcriptomes across many would help to disentangle what is noise and what is signal samples would help address this hypothesis; at this time this would be prohibitively expensive as long-read RNA-seq is approximately five to ten times the cost of short-read RNA-seq.

We used a large dataset compiled from published RNA-seq data to build our ocular transcriptomes, an approach which has several key advantages. First, our large sample size allows us to combat the noisy nature of RNA-seq data. More importantly, because we only keep transcripts detected across multiple samples from multiple studies with multiple types of sample preparation, we can be more confident that our transcriptomes accurately reflect the biology of its originating subtissue and are not a technical artifact due to preparation of the samples.

{== We initially reconstructed `r toString(round(NUM_INITIAL_ANOO_TX/NUM_TOTAL_GENCODE_TX*100), digits = 3)` % of the Gencode transcriptome, but after filtering we were left with `r toString(round(NUM_REF_TX_BODY/NUM_TOTAL_GENCODE_TX*100), digits = 3)` %. This was expected as similar results are seen in the CHESS database with which we share some of our underlying data with. This suggests that some of the Gencode transcripts may not be relevant to every human organ system, and that the one size fits all strategy of having a single comprehensive annotation for all human tissues may not be the most optimal one.(still need to DO SOME LITERATURE SEARCH TO SEE IF THIS IS MENTIONED ANYWHERE. GTEx papers?). This is further supported by the minimal loss loss in gene expression data despite the large decrease in transcriptome size for each tissue. This reduction in total transcriptome sizes is also useful for other researchers as it creates a smaller search space to explore, ie it is easier to compare 30,000 transcripts than 300,000 ==}{>> Maybe cut this paragraph? <<}

In almost all tissues we see more noncoding novel isoforms than novel coding. However as all/most of our data was prepared with polyA selection, we likely are not capturing canonical lincRNA/ncRNA which do not have polyA tails. These transcripts fall under the processed\_transcript category by the Gencode annotation which is a very vague category so it is difficult to predict their function.(find previous work on noncoding polyadenylated transcripts) Many of these have alternative FEs so might be under a different regulatory program. Though we identified many novel ORFs it is difficult to say whether these are actually protein coding. We see that many of the novel ORF's are novel because of retained introns, and these retained intron transcripts may simply be unspliced mRNA. Studies have shown that some neuronal cell types stockpile unspliced transcripts for generating rapid changes in gene expression, and so we may be capturing a similar phenomenon here(CITE). However we are confident that we are capturing true novel protein isoforms because XYZ. We find a lot, but no where near the upper limit suggested by {==  {++ Who et at?? ++}  ==}{>> Add author <<}[-@ponomarenko_size_2016]

We observed that in the set of novel exons within our transcriptome, the vast majority are novel first and last exons [NUMBERS]. It is difficult to directly tell what the biological relevance of these might be. There are multiple studies examining the roles of alternative first and last exons, with first exons arising because of alternative promoter usage, and last exons by alternative polyadenylation. [-@demircioglu_pan-cancer_2019],[-@mitra_alternative_2018] These studies have highlighted the distinct biological role these phenomena play. However, other studies have shown that many of these differences in start sites are largely nonfunctional and lack biological significance.[-@xu_evidence_2019]. Previous research devoted to studying splicing vastly outnumbers other research on other RNA processing types, yet we show that isoforms generated by other RNA mechanisms are equally abundant, and thus merit more study. Something about the reliability of determining novel 5' ends and 3' ends from RNA-seq/de novo transcriptomes, and how we are not capturing all RNA processing mechanisms.

In our ocular transcriptomes, we see that novel isoforms are largely subtissue-specific. This matches previously reported findings about the tissue specificity of rare exons and first and last exons.(cite)We also find that on average novel isoforms represent about 15% of their parent gene's expression. However, it is difficult to say if these are non-functional. Its is entirely possible that these isoforms are from low population cell types. This especially makes sense in the retina, bipolar/amacrine cell types comprise \~15% of the total cell population, and that within these cell types there are even smaller sub cell type populations.(cite) However, we note that transcripts only expressed in rare cell types are likely to be missed by our pipeline, as we used for a conservative expression threshold for the filtering step within our pipeline.

We want to make our transcriptomes easily accessible to other researchers, so we designed a webapp to visualize our transcriptomes and access tissue-specific annotation files. We wanted to provide as much information to the user so that they can draw their own conclusions about the significance of potential novel exon, and so we provide the gene model with novel exon, coding and noncoding regions marked, along with the PIU for each transcript constructed within a gene. We also provide the fraction of samples within a given subtissue type a sample was detected in, to provide a further level of evidence to the validity of constructed transcripts.

The recent GnomAD[-@karczewski_mutational_2020] project showed that taking transcript expression in to account can improve variant prioritization.[-@noauthor_transcript_nodate] Our tissues specfic variant prioitization builds upon this idea by precisely identifying tissue specifc transcript expression to improve expression based variant prioitization.(needs more)

novel loci found some, some are interseting, some are not.

# Supplemental Figures


```{r sup_fig1 }
ggplot(flnc_lengths)+
  geom_violin(aes(x=lib, y=V2))+
  ylab('Transcript Length')+
  xlab('PacBio Library Type')+
  ggtitle('Length Distribution of Full Length Non-Chimeric(FLNC) long reads')+
  cowplot::theme_cowplot()
```

# Notes

-   Correlation between transcript length and expression?

-   bench marking transdecoder on gencode txome.

-   comparison of cage to non cage

# References


